

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>neofox.model.conversion &mdash; Neofox 0.4.0.dev1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link href="../../../_static/style.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home" alt="Documentation Home"> Neofox
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../readme.html">NeoFox - NEOantigen Feature tOolboX</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../models.html">Data models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api_usage.html">Programmatic usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../neofox.html">Neofox package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Neofox</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>neofox.model.conversion</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for neofox.model.conversion</h1><div class="highlight"><pre>
<span></span><span class="c1">#</span>
<span class="c1"># Copyright (c) 2020-2030 Translational Oncology at the Medical Center of the Johannes Gutenberg-University Mainz gGmbH.</span>
<span class="c1">#</span>
<span class="c1"># This file is part of Neofox</span>
<span class="c1"># (see https://github.com/tron-bioinformatics/neofox).</span>
<span class="c1">#</span>
<span class="c1"># This program is free software: you can redistribute it and/or modify</span>
<span class="c1"># it under the terms of the GNU General Public License as published by</span>
<span class="c1"># the Free Software Foundation, either version 3 of the License, or</span>
<span class="c1"># (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1"># This program is distributed in the hope that it will be useful,</span>
<span class="c1"># but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1"># MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1"># GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1"># You should have received a copy of the GNU General Public License</span>
<span class="c1"># along with this program. If not, see &lt;http://www.gnu.org/licenses/&gt;.#</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">betterproto</span>
<span class="kn">from</span> <span class="nn">betterproto</span> <span class="kn">import</span> <span class="n">Casing</span>
<span class="kn">from</span> <span class="nn">pandas.io.json</span> <span class="kn">import</span> <span class="n">json_normalize</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">difflib</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">from</span> <span class="nn">neofox.model.neoantigen</span> <span class="kn">import</span> <span class="n">Neoantigen</span><span class="p">,</span> <span class="n">Transcript</span><span class="p">,</span> <span class="n">Mutation</span><span class="p">,</span> <span class="n">Patient</span><span class="p">,</span> <span class="n">NeoantigenAnnotations</span><span class="p">,</span> <span class="n">Mhc2Name</span><span class="p">,</span> \
    <span class="n">Mhc2GeneName</span><span class="p">,</span> <span class="n">Zygosity</span><span class="p">,</span> <span class="n">Mhc2Gene</span><span class="p">,</span> <span class="n">Mhc2</span><span class="p">,</span> <span class="n">Mhc2Isoform</span><span class="p">,</span> <span class="n">MhcAllele</span><span class="p">,</span> <span class="n">Mhc1Name</span><span class="p">,</span> <span class="n">Mhc1</span>

<span class="n">FIELD_SUBSTITUTION</span> <span class="o">=</span> <span class="s1">&#39;substitution&#39;</span>

<span class="n">FIELD_VAF_DNA</span> <span class="o">=</span> <span class="s1">&#39;VAF_in_tumor&#39;</span>
<span class="n">FIELD_VAF_RNA</span> <span class="o">=</span> <span class="s1">&#39;VAF_in_RNA&#39;</span>
<span class="n">FIELD_RNA_EXPRESSION</span> <span class="o">=</span> <span class="s1">&#39;VAF_RNA_raw&#39;</span>
<span class="n">FIELD_TRANSCRIPT_EXPRESSION</span> <span class="o">=</span> <span class="s1">&#39;transcript_expression&#39;</span>
<span class="n">FIELD_TRANSCRIPT</span> <span class="o">=</span> <span class="s1">&#39;UCSC_transcript&#39;</span>
<span class="n">FIELD_GENE</span> <span class="o">=</span> <span class="s1">&#39;gene&#39;</span>
<span class="n">FIELD_WILD_TYPE_XMER</span> <span class="o">=</span> <span class="s1">&#39;[WT]_+-13_AA_(SNV)_/_-15_AA_to_STOP_(INDEL)&#39;</span>
<span class="n">FIELD_MUTATED_XMER</span> <span class="o">=</span> <span class="s1">&#39;+-13_AA_(SNV)_/_-15_AA_to_STOP_(INDEL)&#39;</span>


<div class="viewcode-block" id="ModelConverter"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter">[docs]</a><span class="k">class</span> <span class="nc">ModelConverter</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>

    <span class="n">HLA_ALLELE_PATTERN</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;(?:HLA-)?([A-Z0-9]+)[\*|_]?([0-9]</span><span class="si">{2}</span><span class="s2">):?([0-9]{2,}):?([0-9]{2,})?:?([0-9]{2,})?([N|L|S|Q]{0,1})&quot;</span><span class="p">)</span>
    <span class="n">HLA_MOLECULE_PATTERN</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(?:HLA-)?([A-Z0-9]+[\*|_]?[0-9]{2,}:?[0-9]{2,})-&quot;</span>
                                      <span class="sa">r</span><span class="s2">&quot;([A-Z0-9]+[\*|_]?[0-9]{2,}:?[0-9]{2,})&quot;</span><span class="p">)</span>
    <span class="n">HLA_DR_MOLECULE_PATTERN</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(?:HLA-)?(DRB1[\*|_]?[0-9]{2,}:?[0-9]{2,})&quot;</span><span class="p">)</span>
    <span class="n">GENES_BY_MOLECULE</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">Mhc2Name</span><span class="o">.</span><span class="n">DR</span><span class="p">:</span> <span class="p">[</span><span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">DRB1</span><span class="p">],</span>
        <span class="n">Mhc2Name</span><span class="o">.</span><span class="n">DP</span><span class="p">:</span> <span class="p">[</span><span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">DPA1</span><span class="p">,</span> <span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">DPB1</span><span class="p">],</span>
        <span class="n">Mhc2Name</span><span class="o">.</span><span class="n">DQ</span><span class="p">:</span> <span class="p">[</span><span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">DQA1</span><span class="p">,</span> <span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">DQB1</span><span class="p">],</span>
    <span class="p">}</span>

<div class="viewcode-block" id="ModelConverter.parse_candidate_file"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.parse_candidate_file">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">parse_candidate_file</span><span class="p">(</span><span class="n">candidate_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">patient_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Neoantigen</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param candidate_file: the path to an neoantigen candidate input file</span>
<span class="sd">        :param patient_id: the patient identifier for all neoantigens in the input file, if not provided it is</span>
<span class="sd">        expected as column named `patient.id` or `patient`</span>
<span class="sd">        :return neoantigens in model objects</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">candidate_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="c1"># filter out indels</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="o">~</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;substitution&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span>
        <span class="n">ModelConverter</span><span class="o">.</span><span class="n">_enrich_candidate_table</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="n">neoantigens</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">candidate_entry</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">neoantigens</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ModelConverter</span><span class="o">.</span><span class="n">_candidate_entry2model</span><span class="p">(</span><span class="n">candidate_entry</span><span class="p">,</span> <span class="n">patient_id</span><span class="o">=</span><span class="n">patient_id</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">neoantigens</span></div>

<div class="viewcode-block" id="ModelConverter.parse_patients_file"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.parse_patients_file">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">parse_patients_file</span><span class="p">(</span><span class="n">patients_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Patient</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param patients_file: the file to patients data CSV file</span>
<span class="sd">        :return: the parsed CSV into model objects</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">split_comma_separated_list</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">patients_file</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
            <span class="n">converters</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;mhcIAlleles&#39;</span><span class="p">:</span> <span class="n">split_comma_separated_list</span><span class="p">,</span>
                        <span class="s1">&#39;mhcIIAlleles&#39;</span><span class="p">:</span> <span class="n">split_comma_separated_list</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">patient_metadata_csv2objects</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></div>

<div class="viewcode-block" id="ModelConverter.parse_neoantigens_file"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.parse_neoantigens_file">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">parse_neoantigens_file</span><span class="p">(</span><span class="n">neoantigens_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Neoantigen</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param neoantigens_file: the file to neoantigens data CSV file</span>
<span class="sd">        :return: the parsed CSV into model objects</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">neoantigens_csv2objects</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">neoantigens_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">))</span></div>

<div class="viewcode-block" id="ModelConverter.objects2dataframe"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.objects2dataframe">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">objects2dataframe</span><span class="p">(</span><span class="n">model_objects</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param model_objects: list of objects of subclass of betterproto.Message</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">json_normalize</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">[</span><span class="n">n</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="n">include_default_values</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">model_objects</span><span class="p">])</span></div>

<div class="viewcode-block" id="ModelConverter.objects2json"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.objects2json">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">objects2json</span><span class="p">(</span><span class="n">model_objects</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">],</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param model_objects: list of objects of subclass of betterproto.Message</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">([</span><span class="n">o</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="n">casing</span><span class="o">=</span><span class="n">Casing</span><span class="o">.</span><span class="n">SNAKE</span><span class="p">)</span> <span class="k">for</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">model_objects</span><span class="p">],</span> <span class="n">f</span><span class="p">)</span></div>

<div class="viewcode-block" id="ModelConverter.object2series"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.object2series">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">object2series</span><span class="p">(</span><span class="n">model_object</span><span class="p">:</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param model_object: object of subclass of betterproto.Message</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">json_normalize</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">model_object</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="n">casing</span><span class="o">=</span><span class="n">Casing</span><span class="o">.</span><span class="n">SNAKE</span><span class="p">,</span> <span class="n">include_default_values</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span></div>

<div class="viewcode-block" id="ModelConverter.object2flat_dict"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.object2flat_dict">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">object2flat_dict</span><span class="p">(</span><span class="n">model</span><span class="p">:</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Transforms a model object into a flat dict. Nested fields are concatenated with a dot</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">object2series</span><span class="p">(</span><span class="n">model</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span></div>

<div class="viewcode-block" id="ModelConverter.neoantigens_csv2object"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.neoantigens_csv2object">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">neoantigens_csv2object</span><span class="p">(</span><span class="n">series</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Neoantigen</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;transforms an entry from a CSV into an object&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">Neoantigen</span><span class="p">()</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">ModelConverter</span><span class="o">.</span><span class="n">_flat_dict2nested_dict</span><span class="p">(</span><span class="n">flat_dict</span><span class="o">=</span><span class="n">series</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()))</span></div>

<div class="viewcode-block" id="ModelConverter.patient_metadata_csv2objects"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.patient_metadata_csv2objects">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">patient_metadata_csv2objects</span><span class="p">(</span><span class="n">dataframe</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Patient</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;transforms an patients CSV into a list of objects&quot;&quot;&quot;</span>
        <span class="n">patients</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">patient_dict</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
            <span class="n">patient</span> <span class="o">=</span> <span class="n">Patient</span><span class="p">()</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">patient_dict</span><span class="p">)</span>
            <span class="n">patient</span><span class="o">.</span><span class="n">mhc1</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">parse_mhc1_alleles</span><span class="p">(</span><span class="n">patient_dict</span><span class="p">[</span><span class="s1">&#39;mhcIAlleles&#39;</span><span class="p">])</span>
            <span class="n">patient</span><span class="o">.</span><span class="n">mhc2</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">parse_mhc2_alleles</span><span class="p">(</span><span class="n">patient_dict</span><span class="p">[</span><span class="s1">&#39;mhcIIAlleles&#39;</span><span class="p">])</span>
            <span class="n">patients</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">patient</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">patients</span></div>

<div class="viewcode-block" id="ModelConverter.neoantigens_csv2objects"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.neoantigens_csv2objects">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">neoantigens_csv2objects</span><span class="p">(</span><span class="n">dataframe</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Neoantigen</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;transforms an patients CSV into a list of objects&quot;&quot;&quot;</span>
        <span class="n">neoantigens</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">neoantigens</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Neoantigen</span><span class="p">()</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">ModelConverter</span><span class="o">.</span><span class="n">_flat_dict2nested_dict</span><span class="p">(</span><span class="n">flat_dict</span><span class="o">=</span><span class="n">row</span><span class="o">.</span><span class="n">to_dict</span><span class="p">())))</span>
        <span class="k">return</span> <span class="n">neoantigens</span></div>

<div class="viewcode-block" id="ModelConverter.annotations2short_wide_table"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.annotations2short_wide_table">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">annotations2short_wide_table</span><span class="p">(</span>
            <span class="n">neoantigen_annotations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">NeoantigenAnnotations</span><span class="p">],</span> <span class="n">neoantigens</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Neoantigen</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="n">dfs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">neoantigens_df</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">objects2dataframe</span><span class="p">(</span><span class="n">neoantigens</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">na</span> <span class="ow">in</span> <span class="n">neoantigen_annotations</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">a</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">na</span><span class="o">.</span><span class="n">annotations</span><span class="p">])</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;identifier&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">na</span><span class="o">.</span><span class="n">neoantigen_identifier</span>
            <span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">del</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;index&#39;</span><span class="p">]</span>
            <span class="n">dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
        <span class="n">annotations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">neoantigens_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;identifier&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">annotations_df</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;identifier&#39;</span><span class="p">)</span></div>

<div class="viewcode-block" id="ModelConverter.annotations2tall_skinny_table"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.annotations2tall_skinny_table">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">annotations2tall_skinny_table</span><span class="p">(</span><span class="n">neoantigen_annotations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">NeoantigenAnnotations</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="n">dfs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">na</span> <span class="ow">in</span> <span class="n">neoantigen_annotations</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">a</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">na</span><span class="o">.</span><span class="n">annotations</span><span class="p">])</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;neoantigen_identifier&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">na</span><span class="o">.</span><span class="n">neoantigen_identifier</span>
            <span class="n">dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">value</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">])</span>    <span class="c1"># avoid writing NA values</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_flat_dict2nested_dict</span><span class="p">(</span><span class="n">flat_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;transforms a flattened dict into a nested dict, assuming that the dot indicates a nested level&quot;&quot;&quot;</span>
        <span class="n">nested_dict</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="p">{})</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">flat_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">splitted_k</span> <span class="o">=</span> <span class="n">k</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">splitted_k</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
                <span class="k">raise</span> <span class="bp">NotImplemented</span><span class="p">(</span><span class="s2">&quot;Support for dictionaries nested more than one level is not implemented&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">splitted_k</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">nested_dict</span><span class="p">[</span><span class="n">splitted_k</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="n">splitted_k</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">v</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">nested_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>
        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">nested_dict</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_candidate_entry2model</span><span class="p">(</span><span class="n">candidate_entry</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">patient_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Neoantigen</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;parses an row from a candidate file into a model object&quot;&quot;&quot;</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="n">Transcript</span><span class="p">()</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">assembly</span> <span class="o">=</span> <span class="s1">&#39;hg19&#39;</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">gene</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">FIELD_GENE</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">FIELD_TRANSCRIPT</span><span class="p">)</span>

        <span class="n">mutation</span> <span class="o">=</span> <span class="n">Mutation</span><span class="p">()</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;position&#39;</span><span class="p">)</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">wild_type_aminoacid</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;wild_type_aminoacid&#39;</span><span class="p">)</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">mutated_aminoacid</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;mutated_aminoacid&#39;</span><span class="p">)</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">left_flanking_region</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;left_flanking_region&#39;</span><span class="p">)</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">right_flanking_region</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;right_flanking_region&#39;</span><span class="p">)</span>

        <span class="n">neoantigen</span> <span class="o">=</span> <span class="n">Neoantigen</span><span class="p">()</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">patient_identifier</span> <span class="o">=</span> <span class="n">patient_id</span> <span class="k">if</span> <span class="n">patient_id</span> <span class="k">else</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;patient&#39;</span><span class="p">)</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span> <span class="o">=</span> <span class="n">mutation</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">transcript</span> <span class="o">=</span> <span class="n">transcript</span>
        <span class="c1"># clonality estimation is not present in candidate file at the moment</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">clonality_estimation</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1"># missing RNA expression values are represented as -1</span>
        <span class="n">vaf_rna_raw</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">FIELD_TRANSCRIPT_EXPRESSION</span><span class="p">)</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">rna_expression</span> <span class="o">=</span> <span class="n">vaf_rna_raw</span> <span class="k">if</span> <span class="n">vaf_rna_raw</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="n">vaf_in_rna</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">FIELD_VAF_RNA</span><span class="p">)</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">rna_variant_allele_frequency</span> <span class="o">=</span> <span class="n">vaf_in_rna</span> <span class="k">if</span> <span class="n">vaf_in_rna</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">dna_variant_allele_frequency</span> <span class="o">=</span> <span class="n">candidate_entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">FIELD_VAF_DNA</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">neoantigen</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_enrich_candidate_table</span><span class="p">(</span><span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;parses some data from the candidate table into the right fields in the model objects&quot;&quot;&quot;</span>
        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;wild_type_aminoacid&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">FIELD_SUBSTITUTION</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;(\w)\d+\w&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;mutated_aminoacid&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">FIELD_SUBSTITUTION</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;\w\d+(\w)&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">FIELD_SUBSTITUTION</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;\w(\d+)\w&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)))</span>
        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;left_flanking_region&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[[</span>
            <span class="n">FIELD_MUTATED_XMER</span><span class="p">,</span> <span class="n">FIELD_WILD_TYPE_XMER</span><span class="p">]]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">_get_matching_region</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;right_flanking_region&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[[</span>
            <span class="n">FIELD_MUTATED_XMER</span><span class="p">,</span> <span class="n">FIELD_WILD_TYPE_XMER</span><span class="p">]]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">_get_matching_region</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">match</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_get_matching_region</span><span class="p">(</span><span class="n">sequence1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sequence2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">match</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;fetches an aligned block within two sequences&quot;&quot;&quot;</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">difflib</span><span class="o">.</span><span class="n">SequenceMatcher</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">sequence1</span><span class="p">,</span> <span class="n">sequence2</span><span class="p">)</span><span class="o">.</span><span class="n">get_matching_blocks</span><span class="p">()[</span><span class="n">match</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">sequence1</span><span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">a</span> <span class="p">:</span> <span class="n">match</span><span class="o">.</span><span class="n">a</span> <span class="o">+</span> <span class="n">match</span><span class="o">.</span><span class="n">size</span><span class="p">]</span>

<div class="viewcode-block" id="ModelConverter.parse_mhc1_alleles"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.parse_mhc1_alleles">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">parse_mhc1_alleles</span><span class="p">(</span><span class="n">alleles</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Mhc1</span><span class="p">]:</span>
        <span class="n">isoforms</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">parsed_alleles</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">ModelConverter</span><span class="o">.</span><span class="n">parse_mhc_allele</span><span class="p">,</span> <span class="n">alleles</span><span class="p">))</span>
        <span class="n">ModelConverter</span><span class="o">.</span><span class="n">_validate_mhc1_alleles</span><span class="p">(</span><span class="n">parsed_alleles</span><span class="p">)</span>
        <span class="c1"># do we need to validate genes anymore? add test creating MhcAllele with bad gene and see what happens</span>
        <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">Mhc1Name</span><span class="p">:</span>
            <span class="n">gene_alleles</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">a</span><span class="p">:</span> <span class="n">a</span><span class="o">.</span><span class="n">gene</span> <span class="o">==</span> <span class="n">gene_name</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">parsed_alleles</span><span class="p">))</span>
            <span class="n">zygosity</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">_get_zygosity_from_alleles</span><span class="p">(</span><span class="n">gene_alleles</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">zygosity</span> <span class="o">==</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">HOMOZYGOUS</span><span class="p">:</span>
                <span class="n">gene_alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene_alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>   <span class="c1"># we don&#39;t want repeated instances of the same allele</span>
            <span class="n">isoforms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Mhc1</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">zygosity</span><span class="o">=</span><span class="n">zygosity</span><span class="p">,</span> <span class="n">alleles</span><span class="o">=</span><span class="n">gene_alleles</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">isoforms</span></div>

<div class="viewcode-block" id="ModelConverter.parse_mhc2_alleles"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.parse_mhc2_alleles">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">parse_mhc2_alleles</span><span class="p">(</span><span class="n">alleles</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Mhc2</span><span class="p">]:</span>
        <span class="n">mhc2s</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">parsed_alleles</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">ModelConverter</span><span class="o">.</span><span class="n">parse_mhc_allele</span><span class="p">,</span> <span class="n">alleles</span><span class="p">))</span>
        <span class="n">ModelConverter</span><span class="o">.</span><span class="n">_validate_mhc2_alleles</span><span class="p">(</span><span class="n">parsed_alleles</span><span class="p">)</span>
        <span class="c1"># do we need to validate genes anymore? add test creating MhcAllele with bad gene and see what happens</span>
        <span class="k">for</span> <span class="n">isoform_name</span> <span class="ow">in</span> <span class="n">Mhc2Name</span><span class="p">:</span>
            <span class="n">isoform_alleles</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">a</span><span class="p">:</span> <span class="n">isoform_name</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">a</span><span class="o">.</span><span class="n">gene</span><span class="p">,</span> <span class="n">parsed_alleles</span><span class="p">))</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">GENES_BY_MOLECULE</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">isoform_name</span><span class="p">):</span>
                <span class="n">gene_alleles</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">a</span><span class="p">:</span> <span class="n">a</span><span class="o">.</span><span class="n">gene</span> <span class="o">==</span> <span class="n">gene_name</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">isoform_alleles</span><span class="p">))</span>
                <span class="n">zygosity</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">_get_zygosity_from_alleles</span><span class="p">(</span><span class="n">gene_alleles</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">zygosity</span> <span class="o">==</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">HOMOZYGOUS</span><span class="p">:</span>
                    <span class="n">gene_alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene_alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>  <span class="c1"># we don&#39;t want repeated instances of the same allele</span>
                <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Mhc2Gene</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">zygosity</span><span class="o">=</span><span class="n">zygosity</span><span class="p">,</span> <span class="n">alleles</span><span class="o">=</span><span class="n">gene_alleles</span><span class="p">))</span>
            <span class="n">isoforms</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">_get_mhc2_isoforms</span><span class="p">(</span><span class="n">isoform_name</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span>
            <span class="n">mhc2s</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Mhc2</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">isoform_name</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span> <span class="n">isoforms</span><span class="o">=</span><span class="n">isoforms</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">mhc2s</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_get_mhc2_isoforms</span><span class="p">(</span><span class="n">isoform_name</span><span class="p">:</span> <span class="n">Mhc2Name</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Mhc2Gene</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Mhc2Isoform</span><span class="p">]:</span>
        <span class="n">isoforms</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">isoform_name</span> <span class="o">==</span> <span class="n">Mhc2Name</span><span class="o">.</span><span class="n">DR</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;More than one gene provided for MHC II DR&quot;</span>
            <span class="c1"># alpha chain of the MHC II DR is not modelled as it is constant</span>
            <span class="n">isoforms</span> <span class="o">=</span> <span class="p">[</span><span class="n">Mhc2Isoform</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">alpha_chain</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">beta_chain</span><span class="o">=</span><span class="n">a</span><span class="p">)</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">alleles</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">isoform_name</span> <span class="o">==</span> <span class="n">Mhc2Name</span><span class="o">.</span><span class="n">DP</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;More than two genes provided for MHC II DP&quot;</span>
            <span class="n">alpha_alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">DPA1</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">alleles</span><span class="p">]</span>
            <span class="n">beta_alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">DPB1</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">alleles</span><span class="p">]</span>
            <span class="n">isoforms</span> <span class="o">=</span> <span class="p">[</span><span class="n">Mhc2Isoform</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">ModelConverter</span><span class="o">.</span><span class="n">get_mhc2_isoform_name</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">),</span>
                                        <span class="n">alpha_chain</span><span class="o">=</span><span class="n">a</span><span class="p">,</span> <span class="n">beta_chain</span><span class="o">=</span><span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">alpha_alleles</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">beta_alleles</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">isoform_name</span> <span class="o">==</span> <span class="n">Mhc2Name</span><span class="o">.</span><span class="n">DQ</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;More than two genes provided for MHC II DQ&quot;</span>
            <span class="n">alpha_alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">DQA1</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">alleles</span><span class="p">]</span>
            <span class="n">beta_alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">DQB1</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">alleles</span><span class="p">]</span>
            <span class="n">isoforms</span> <span class="o">=</span> <span class="p">[</span><span class="n">Mhc2Isoform</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">ModelConverter</span><span class="o">.</span><span class="n">get_mhc2_isoform_name</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">),</span>
                                        <span class="n">alpha_chain</span><span class="o">=</span><span class="n">a</span><span class="p">,</span> <span class="n">beta_chain</span><span class="o">=</span><span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">alpha_alleles</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">beta_alleles</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">isoforms</span>

<div class="viewcode-block" id="ModelConverter.get_mhc2_isoform_name"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.get_mhc2_isoform_name">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">get_mhc2_isoform_name</span><span class="p">(</span><span class="n">a</span><span class="p">:</span> <span class="n">MhcAllele</span><span class="p">,</span> <span class="n">b</span><span class="p">:</span> <span class="n">MhcAllele</span><span class="p">):</span>
        <span class="c1"># NOTE: this is needed as jus setting alpha chain to None wouldn&#39;t work with protobuf</span>
        <span class="k">if</span> <span class="n">a</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">b</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;HLA-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">b</span><span class="o">.</span><span class="n">name</span></div>

<div class="viewcode-block" id="ModelConverter.parse_mhc_allele"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.conversion.ModelConverter.parse_mhc_allele">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">parse_mhc_allele</span><span class="p">(</span><span class="n">allele</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MhcAllele</span><span class="p">:</span>
        <span class="c1"># infers gene, group and protein from the name</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">HLA_ALLELE_PATTERN</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">allele</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Allele does not match HLA allele pattern </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">allele</span><span class="p">)</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">protein</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
        <span class="c1"># builds a normalized representation of the allele</span>
        <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;HLA-</span><span class="si">{gene}</span><span class="s2">*</span><span class="si">{serotype}</span><span class="s2">:</span><span class="si">{protein}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span> <span class="n">serotype</span><span class="o">=</span><span class="n">group</span><span class="p">,</span> <span class="n">protein</span><span class="o">=</span><span class="n">protein</span><span class="p">)</span>
        <span class="c1"># ensures that full name stores the complete allele as provided but normalizes</span>
        <span class="c1"># its representation</span>
        <span class="n">full_name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="n">six_digits_id</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">six_digits_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">six_digits_id</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">full_name</span> <span class="o">=</span> <span class="n">full_name</span> <span class="o">+</span> <span class="s2">&quot;:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">six_digits_id</span><span class="p">)</span>
            <span class="n">eight_digits_id</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">eight_digits_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">eight_digits_id</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">full_name</span> <span class="o">=</span> <span class="n">full_name</span> <span class="o">+</span> <span class="s2">&quot;:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">eight_digits_id</span><span class="p">)</span>
                <span class="n">expression_change</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">expression_change</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">expression_change</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="n">full_name</span> <span class="o">=</span> <span class="n">full_name</span> <span class="o">+</span> <span class="n">expression_change</span>
        <span class="k">return</span> <span class="n">MhcAllele</span><span class="p">(</span><span class="n">full_name</span><span class="o">=</span><span class="n">full_name</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">gene</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">group</span><span class="p">,</span> <span class="n">protein</span><span class="o">=</span><span class="n">protein</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_get_zygosity_from_alleles</span><span class="p">(</span><span class="n">alleles</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">MhcAllele</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Zygosity</span><span class="p">:</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">a</span><span class="o">.</span><span class="n">gene</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">]))</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;Trying to get zygosity from alleles of different genes&quot;</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;More than 2 alleles for gene </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
            <span class="n">zygosity</span> <span class="o">=</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">HOMOZYGOUS</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">zygosity</span> <span class="o">=</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">HETEROZYGOUS</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">zygosity</span> <span class="o">=</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">HEMIZYGOUS</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">zygosity</span> <span class="o">=</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">LOSS</span>
        <span class="k">return</span> <span class="n">zygosity</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_validate_mhc1_alleles</span><span class="p">(</span><span class="n">parsed_alleles</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">MhcAllele</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">parsed_alleles</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">a</span><span class="o">.</span><span class="n">gene</span> <span class="ow">in</span> <span class="n">Mhc1Name</span><span class="o">.</span><span class="n">__members__</span><span class="p">,</span> <span class="s2">&quot;Gene from MHC I allele is not valid: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_validate_mhc2_alleles</span><span class="p">(</span><span class="n">parsed_alleles</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">MhcAllele</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">parsed_alleles</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">a</span><span class="o">.</span><span class="n">gene</span> <span class="ow">in</span> <span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">__members__</span><span class="p">,</span> <span class="s2">&quot;Gene from MHC II allele is not valid: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Franziska Lang &amp; Pablo Riesgo Ferreiro

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>