

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>neofox.model.neoantigen &mdash; Neofox 0.4.0.dev1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link href="../../../_static/style.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home" alt="Documentation Home"> Neofox
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../readme.html">NeoFox - NEOantigen Feature tOolboX</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../models.html">Data models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api_usage.html">Programmatic usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../neofox.html">Neofox package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Neofox</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>neofox.model.neoantigen</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for neofox.model.neoantigen</h1><div class="highlight"><pre>
<span></span><span class="c1"># Generated by the protocol buffer compiler.  DO NOT EDIT!</span>
<span class="c1"># sources: neoantigen.proto</span>
<span class="c1"># plugin: python-betterproto</span>
<span class="kn">from</span> <span class="nn">dataclasses</span> <span class="kn">import</span> <span class="n">dataclass</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span>

<span class="kn">import</span> <span class="nn">betterproto</span>


<div class="viewcode-block" id="Zygosity"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Zygosity">[docs]</a><span class="k">class</span> <span class="nc">Zygosity</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Enum</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;*The zygosity of a given gene&quot;&quot;&quot;</span>

    <span class="c1"># *Two equal copies of the gene</span>
    <span class="n">HOMOZYGOUS</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># *Two different copies of the gene</span>
    <span class="n">HETEROZYGOUS</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="c1"># *Only one copy of the gene</span>
    <span class="n">HEMIZYGOUS</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="c1"># *No copy of the gene</span>
    <span class="n">LOSS</span> <span class="o">=</span> <span class="mi">3</span></div>


<div class="viewcode-block" id="Mhc1Name"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Mhc1Name">[docs]</a><span class="k">class</span> <span class="nc">Mhc1Name</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Enum</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;*Valid names for MHC I classic genes&quot;&quot;&quot;</span>

    <span class="n">A</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">B</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">C</span> <span class="o">=</span> <span class="mi">2</span></div>


<div class="viewcode-block" id="Mhc2GeneName"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Mhc2GeneName">[docs]</a><span class="k">class</span> <span class="nc">Mhc2GeneName</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Enum</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    *Valid names for MHC II classic genes.DRA is not included in this list as</span>
<span class="sd">    it does not have much variability in the population and for our purpose</span>
<span class="sd">    isconsidered constant.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">DRB1</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">DPA1</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">DPB1</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">DQA1</span> <span class="o">=</span> <span class="mi">3</span>
    <span class="n">DQB1</span> <span class="o">=</span> <span class="mi">4</span></div>


<div class="viewcode-block" id="Mhc2Name"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Mhc2Name">[docs]</a><span class="k">class</span> <span class="nc">Mhc2Name</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Enum</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;*Valid names for MHC II classic molecules&quot;&quot;&quot;</span>

    <span class="n">DR</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">DP</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">DQ</span> <span class="o">=</span> <span class="mi">2</span></div>


<div class="viewcode-block" id="Transcript"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Transcript">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">Transcript</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="c1"># *The transcript identifier to which this neoepitope definition refers</span>
    <span class="c1"># (e.g.: Ensembl transcript id)</span>
    <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *The genome assembly to which the gene definition refers to (e.g.: GRCh37,</span>
    <span class="c1"># GRCh38)</span>
    <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *The gene symbol or gene identifier, optional as the transcript</span>
    <span class="c1"># unequivocally identifies a gene</span>
    <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span></div>


<div class="viewcode-block" id="Mutation"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Mutation">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">Mutation</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="c1"># *The aminoacid position within the protein. 1-based, starting in the</span>
    <span class="c1"># N-terminus</span>
    <span class="n">position</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">int32_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *Sequence of aminoacids for the wild type xmer</span>
    <span class="n">wild_type_xmer</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *IUPAC code for the wild type aminoacid in this position</span>
    <span class="n">wild_type_aminoacid</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
    <span class="c1"># *Sequence of aminoacids for the mutated xmer</span>
    <span class="n">mutated_xmer</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
    <span class="c1"># *IUPAC code for the mutated aminoacid in this position</span>
    <span class="n">mutated_aminoacid</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    <span class="c1"># *The left flanking region of variable size in IUPAC codes</span>
    <span class="n">left_flanking_region</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span>
    <span class="c1"># *The size of the left flanking region</span>
    <span class="n">size_left_flanking_region</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">int32_field</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
    <span class="c1"># *The right flanking region of variable size in IUPAC codes</span>
    <span class="n">right_flanking_region</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
    <span class="c1"># *The size of the right flanking region</span>
    <span class="n">size_right_flanking_region</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">int32_field</span><span class="p">(</span><span class="mi">9</span><span class="p">)</span></div>


<div class="viewcode-block" id="Neoantigen"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Neoantigen">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">Neoantigen</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;*A neoantigen minimal definition&quot;&quot;&quot;</span>

    <span class="c1"># *A unique identifier of a neoantigen</span>
    <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *Patient identifier</span>
    <span class="n">patient_identifier</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *The transcript where the neoepitope was observed</span>
    <span class="n">transcript</span><span class="p">:</span> <span class="s2">&quot;Transcript&quot;</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
    <span class="c1"># *The mutation</span>
    <span class="n">mutation</span><span class="p">:</span> <span class="s2">&quot;Mutation&quot;</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
    <span class="c1"># *Clonality estimation. At the moment this is a boolean indicating whether</span>
    <span class="c1"># there is clonality or not,there is no quantitive measurement at the moment.</span>
    <span class="n">clonality_estimation</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">bool_field</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    <span class="c1"># *Expression value of the transcript from RNA data. Range [0, +inf].</span>
    <span class="n">rna_expression</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">float_field</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span>
    <span class="c1"># *Variant allele frequency from the DNA. Range [0.0, 1.0]</span>
    <span class="n">dna_variant_allele_frequency</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">float_field</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
    <span class="c1"># *Variant allele frequency from the RNA. Range [0.0, 1.0]</span>
    <span class="n">rna_variant_allele_frequency</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">float_field</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span></div>


<div class="viewcode-block" id="Patient"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Patient">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">Patient</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    *The metadata required for analysis for a given patient + its patient</span>
<span class="sd">    identifier</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># *Patient identifier</span>
    <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *Is RNA expression available?</span>
    <span class="n">is_rna_available</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">bool_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *MHC I classic molecules</span>
    <span class="n">mhc1</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="s2">&quot;Mhc1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
    <span class="c1"># *MHC II classic molecules</span>
    <span class="n">mhc2</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="s2">&quot;Mhc2&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span></div>


<div class="viewcode-block" id="Annotation"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Annotation">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">Annotation</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;*This is a generic class to hold annotations from Neofox&quot;&quot;&quot;</span>

    <span class="c1"># *The name of the annotation</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *The value of the annotationTODO: make this better to hold int, float and</span>
    <span class="c1"># bool</span>
    <span class="n">value</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span></div>


<div class="viewcode-block" id="NeoantigenAnnotations"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.NeoantigenAnnotations">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">NeoantigenAnnotations</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;*A set of annotations for a neoantigen&quot;&quot;&quot;</span>

    <span class="c1"># *A unique identifier of a neoantigen</span>
    <span class="n">neoantigen_identifier</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *List of annotations</span>
    <span class="n">annotations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="s2">&quot;Annotation&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *The annotator</span>
    <span class="n">annotator</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
    <span class="c1"># *The version of the annotator</span>
    <span class="n">annotator_version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
    <span class="c1"># *A timestamp determined when the annotation was created</span>
    <span class="n">timestamp</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    <span class="c1"># *Annotation resources MD5 hash</span>
    <span class="n">resources_hash</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span></div>


<div class="viewcode-block" id="Mhc1"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Mhc1">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">Mhc1</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    *Models MHC I alleles related to the same MHC I gene, i.e. 2 alleles/2</span>
<span class="sd">    isoforms per gene</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># *MHC I gene name</span>
    <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;Mhc1Name&quot;</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">enum_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *Zygosity of the gene</span>
    <span class="n">zygosity</span><span class="p">:</span> <span class="s2">&quot;Zygosity&quot;</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">enum_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *The alleles of the gene (0, 1 or 2)</span>
    <span class="n">alleles</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="s2">&quot;MhcAllele&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span></div>


<div class="viewcode-block" id="Mhc2"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Mhc2">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">Mhc2</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    *Models MHC II alleles related to the same MHC II protein, i.e. 4 isoforms</span>
<span class="sd">    related to 2 genes with 2 alleles each</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># *MHC II molecule name</span>
    <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;Mhc2Name&quot;</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">enum_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *List of MHC II genes</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="s2">&quot;Mhc2Gene&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *Different combinations of MHC II alleles building different isoforms</span>
    <span class="n">isoforms</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="s2">&quot;Mhc2Isoform&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span></div>


<div class="viewcode-block" id="Mhc2Isoform"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Mhc2Isoform">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">Mhc2Isoform</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;*MHC II isoform&quot;&quot;&quot;</span>

    <span class="c1"># *Name to refer to the MHC II isoform</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *The alpha chain of the isoform</span>
    <span class="n">alpha_chain</span><span class="p">:</span> <span class="s2">&quot;MhcAllele&quot;</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *The beta chain of the isoform</span>
    <span class="n">beta_chain</span><span class="p">:</span> <span class="s2">&quot;MhcAllele&quot;</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span></div>


<div class="viewcode-block" id="Mhc2Gene"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.Mhc2Gene">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">Mhc2Gene</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;*MHC II gene&quot;&quot;&quot;</span>

    <span class="c1"># *MHC II gene name</span>
    <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;Mhc2GeneName&quot;</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">enum_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *Zygosity of the gene</span>
    <span class="n">zygosity</span><span class="p">:</span> <span class="s2">&quot;Zygosity&quot;</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">enum_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *The alleles of the gene (0, 1 or 2)</span>
    <span class="n">alleles</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="s2">&quot;MhcAllele&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">message_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span></div>


<div class="viewcode-block" id="MhcAllele"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.neoantigen.MhcAllele">[docs]</a><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">MhcAllele</span><span class="p">(</span><span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    *MHC allele representation. It does not include non synonymous changes to</span>
<span class="sd">    the sequence, changes in the non coding regionor changes in expression. See</span>
<span class="sd">    http://hla.alleles.org/nomenclature/naming.html for details</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># *HLA full name as provided by the user (e.g.: HLA-DRB1*13:01:02:03N). This</span>
    <span class="c1"># will be parsed into name, gene and group.Any digit format is allowed for</span>
    <span class="c1"># this field (ie: 4, 6 or 8 digits), 2 digits names are not specific enough</span>
    <span class="c1"># for ourpurpose and thus invalid</span>
    <span class="n">full_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># *A specific HLA protein (e.g. HLA-DRB1*13:01). Alleles whose numbers differ</span>
    <span class="c1"># in group and protein must differ in oneor more nucleotide substitutions</span>
    <span class="c1"># that change the amino acid sequence of the encoded protein.This name is</span>
    <span class="c1"># normalized to avoid different representations of the same allele. For</span>
    <span class="c1"># instance both HLA-DRB113:01 andHLA-DRB1*13:01:02:03N will be transformed</span>
    <span class="c1"># into their normalised version HLA-DRB1*13:01. This name is also truncatedto</span>
    <span class="c1"># 4 digits. 2 digits names are not specific enough for our purpose and thus</span>
    <span class="c1"># invalid</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># *The gene from either MHC I or II (e.g. DRB1, A) (this information is</span>
    <span class="c1"># redundant with the Mhc1Gene.name andMhc2Gene.name but it is convenient to</span>
    <span class="c1"># have this at this level too, code will check for data coherence)</span>
    <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
    <span class="c1"># *A group of alleles defined by a common serotype ie: Serological antigen</span>
    <span class="c1"># carried by an allotype (e.g. 13 from HLA-DRB1*13)</span>
    <span class="n">group</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
    <span class="c1"># *A specific protein (e.g.: 02 from HLA-DRB1*13:02)</span>
    <span class="n">protein</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">string_field</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Franziska Lang &amp; Pablo Riesgo Ferreiro

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>