

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>neofox.model.validation &mdash; Neofox 0.4.0.dev1 documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link href="../../../_static/style.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home" alt="Documentation Home"> Neofox
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../readme.html">NeoFox - NEOantigen Feature tOolboX</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../models.html">Data models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api_usage.html">Programmatic usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../neofox.html">Neofox package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Neofox</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>neofox.model.validation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for neofox.model.validation</h1><div class="highlight"><pre>
<span></span><span class="c1">#</span>
<span class="c1"># Copyright (c) 2020-2030 Translational Oncology at the Medical Center of the Johannes Gutenberg-University Mainz gGmbH.</span>
<span class="c1">#</span>
<span class="c1"># This file is part of Neofox</span>
<span class="c1"># (see https://github.com/tron-bioinformatics/neofox).</span>
<span class="c1">#</span>
<span class="c1"># This program is free software: you can redistribute it and/or modify</span>
<span class="c1"># it under the terms of the GNU General Public License as published by</span>
<span class="c1"># the Free Software Foundation, either version 3 of the License, or</span>
<span class="c1"># (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1"># This program is distributed in the hope that it will be useful,</span>
<span class="c1"># but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1"># MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1"># GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1"># You should have received a copy of the GNU General Public License</span>
<span class="c1"># along with this program. If not, see &lt;http://www.gnu.org/licenses/&gt;.#</span>
<span class="kn">import</span> <span class="nn">base64</span>
<span class="kn">import</span> <span class="nn">hashlib</span>
<span class="kn">import</span> <span class="nn">betterproto</span>
<span class="kn">from</span> <span class="nn">neofox.model.conversion</span> <span class="kn">import</span> <span class="n">ModelConverter</span>
<span class="kn">from</span> <span class="nn">neofox.exceptions</span> <span class="kn">import</span> <span class="n">NeofoxDataValidationException</span>
<span class="kn">from</span> <span class="nn">neofox.model.neoantigen</span> <span class="kn">import</span> <span class="n">Neoantigen</span><span class="p">,</span> <span class="n">Mutation</span><span class="p">,</span> <span class="n">Transcript</span><span class="p">,</span> <span class="n">Patient</span><span class="p">,</span> <span class="n">MhcAllele</span><span class="p">,</span> <span class="n">Mhc2Isoform</span><span class="p">,</span> <span class="n">Mhc1</span><span class="p">,</span> \
    <span class="n">Mhc1Name</span><span class="p">,</span> <span class="n">Zygosity</span><span class="p">,</span> <span class="n">Mhc2</span><span class="p">,</span> <span class="n">Mhc2Name</span><span class="p">,</span> <span class="n">Mhc2GeneName</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet.IUPAC</span> <span class="kn">import</span> <span class="n">ExtendedIUPACProtein</span><span class="p">,</span> <span class="n">IUPACData</span>


<div class="viewcode-block" id="ModelValidator"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.validation.ModelValidator">[docs]</a><span class="k">class</span> <span class="nc">ModelValidator</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>

<div class="viewcode-block" id="ModelValidator.validate"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.validation.ModelValidator.validate">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">validate</span><span class="p">(</span><span class="n">model</span><span class="p">:</span> <span class="n">betterproto</span><span class="o">.</span><span class="n">Message</span><span class="p">):</span>
        <span class="c1"># TODO: make this method capture appropriately validation issues whend dealing with int and float</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">model</span><span class="o">.</span><span class="fm">__bytes__</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">NeofoxDataValidationException</span><span class="p">(</span><span class="n">e</span><span class="p">)</span></div>

    <span class="c1"># TODO: add patient validation: validate GTEx tissue and MHC alleles</span>

<div class="viewcode-block" id="ModelValidator.validate_neoantigen"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.validation.ModelValidator.validate_neoantigen">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">validate_neoantigen</span><span class="p">(</span><span class="n">neoantigen</span><span class="p">:</span> <span class="n">Neoantigen</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Neoantigen</span><span class="p">:</span>

        <span class="c1"># checks format consistency first</span>
        <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate</span><span class="p">(</span><span class="n">neoantigen</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># checks gene</span>
            <span class="c1"># TODO: do we want to verify existence of gene and transcript id?</span>
            <span class="n">neoantigen</span><span class="o">.</span><span class="n">transcript</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_transcript</span><span class="p">(</span><span class="n">neoantigen</span><span class="o">.</span><span class="n">transcript</span><span class="p">)</span>

            <span class="c1"># checks mutation</span>
            <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_mutation</span><span class="p">(</span><span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="p">)</span>

            <span class="c1"># check the expression values</span>
            <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_expression_values</span><span class="p">(</span><span class="n">neoantigen</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">AssertionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">NeofoxDataValidationException</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>

        <span class="c1"># infer other fields from the model</span>
        <span class="k">return</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_enrich_neoantigen</span><span class="p">(</span><span class="n">neoantigen</span><span class="p">)</span></div>

<div class="viewcode-block" id="ModelValidator.validate_patient"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.validation.ModelValidator.validate_patient">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">validate_patient</span><span class="p">(</span><span class="n">patient</span><span class="p">:</span> <span class="n">Patient</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Patient</span><span class="p">:</span>

        <span class="c1"># checks format consistency first</span>
        <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate</span><span class="p">(</span><span class="n">patient</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># checks that patient id is not empty considering white spaces</span>
            <span class="n">patient_id</span> <span class="o">=</span> <span class="n">patient</span><span class="o">.</span><span class="n">identifier</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">if</span> <span class="n">patient</span><span class="o">.</span><span class="n">identifier</span> <span class="k">else</span> <span class="n">patient</span><span class="o">.</span><span class="n">identifier</span>
            <span class="k">assert</span> <span class="n">patient_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">patient_id</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;Patient identifier is empty&quot;</span>
            <span class="n">patient</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="n">patient_id</span>

            <span class="c1"># TODO: validate new model with isoforms, genes and alleles</span>
            <span class="c1"># checks MHC I</span>
            <span class="n">validated_mhc1s</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">patient</span><span class="o">.</span><span class="n">mhc1</span><span class="p">:</span>
                <span class="n">validated_mhc1s</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_mhc1</span><span class="p">(</span><span class="n">m</span><span class="p">))</span>
            <span class="n">patient</span><span class="o">.</span><span class="n">mhc1</span> <span class="o">=</span> <span class="n">validated_mhc1s</span>
            <span class="c1"># checks MHC II</span>
            <span class="n">validated_mhc2s</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">patient</span><span class="o">.</span><span class="n">mhc2</span><span class="p">:</span>
                <span class="n">validated_mhc2s</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_mhc2</span><span class="p">(</span><span class="n">m</span><span class="p">))</span>
            <span class="n">patient</span><span class="o">.</span><span class="n">mhc2</span> <span class="o">=</span> <span class="n">validated_mhc2s</span>

        <span class="k">except</span> <span class="ne">AssertionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">NeofoxDataValidationException</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">patient</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_validate_mhc1</span><span class="p">(</span><span class="n">mhc1</span><span class="p">:</span> <span class="n">Mhc1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Mhc1</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">mhc1</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">Mhc1Name</span><span class="p">,</span> <span class="s2">&quot;Invalid MHC I name&quot;</span>
        <span class="k">assert</span> <span class="n">mhc1</span><span class="o">.</span><span class="n">zygosity</span> <span class="ow">in</span> <span class="n">Zygosity</span><span class="p">,</span> <span class="s2">&quot;Invalid zygosity&quot;</span>
        <span class="n">alleles</span> <span class="o">=</span> <span class="n">mhc1</span><span class="o">.</span><span class="n">alleles</span>
        <span class="k">if</span> <span class="n">mhc1</span><span class="o">.</span><span class="n">zygosity</span> <span class="ow">in</span> <span class="p">[</span><span class="n">Zygosity</span><span class="o">.</span><span class="n">HOMOZYGOUS</span><span class="p">,</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">HEMIZYGOUS</span><span class="p">]:</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;An homozygous gene must have 1 allele and not </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">mhc1</span><span class="o">.</span><span class="n">zygosity</span> <span class="o">==</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">HETEROZYGOUS</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;An heterozygous or hemizygous gene must have 2 alleles and not </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">mhc1</span><span class="o">.</span><span class="n">zygosity</span> <span class="o">==</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">LOSS</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;An lost gene must have 0 alleles and not </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))</span>
        <span class="n">validated_alleles</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
            <span class="n">validated_allele</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate_mhc_allele_representation</span><span class="p">(</span><span class="n">allele</span><span class="p">)</span>
            <span class="n">validated_alleles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">validated_allele</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">validated_allele</span><span class="o">.</span><span class="n">gene</span> <span class="o">==</span> <span class="n">mhc1</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> \
                <span class="s2">&quot;The allele referring to gene </span><span class="si">{}</span><span class="s2"> is inside gene </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">validated_allele</span><span class="o">.</span><span class="n">gene</span><span class="p">,</span> <span class="n">mhc1</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">mhc1</span><span class="o">.</span><span class="n">alleles</span> <span class="o">=</span> <span class="n">validated_alleles</span>
        <span class="k">return</span> <span class="n">mhc1</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_validate_mhc2</span><span class="p">(</span><span class="n">mhc2</span><span class="p">:</span> <span class="n">Mhc2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Mhc2</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">mhc2</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">Mhc2Name</span><span class="p">,</span> <span class="s2">&quot;Invalid MHC II name&quot;</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">mhc2</span><span class="o">.</span><span class="n">genes</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">gene</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">Mhc2GeneName</span><span class="p">,</span> <span class="s2">&quot;Invalid gene name from MHC II&quot;</span>
            <span class="k">assert</span> <span class="n">gene</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">GENES_BY_MOLECULE</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mhc2</span><span class="o">.</span><span class="n">name</span><span class="p">),</span> \
                <span class="s2">&quot;Gene </span><span class="si">{}</span><span class="s2"> referring to isoform </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">mhc2</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">gene</span><span class="o">.</span><span class="n">zygosity</span> <span class="ow">in</span> <span class="n">Zygosity</span><span class="p">,</span> <span class="s2">&quot;Invalid zygosity&quot;</span>
            <span class="n">alleles</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">alleles</span>
            <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">zygosity</span> <span class="o">==</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">HOMOZYGOUS</span><span class="p">:</span>
                <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;An homozygous gene must have 1 allele and not </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))</span>
            <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">zygosity</span> <span class="ow">in</span> <span class="p">[</span><span class="n">Zygosity</span><span class="o">.</span><span class="n">HETEROZYGOUS</span><span class="p">,</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">HEMIZYGOUS</span><span class="p">]:</span>
                <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;An heterozygous or hemizygous gene must have 2 alleles and not </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))</span>
            <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">zygosity</span> <span class="o">==</span> <span class="n">Zygosity</span><span class="o">.</span><span class="n">LOSS</span><span class="p">:</span>
                <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;An lost gene must have 0 alleles and not </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))</span>
            <span class="n">validated_alleles</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
                <span class="n">validated_allele</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate_mhc_allele_representation</span><span class="p">(</span><span class="n">allele</span><span class="p">)</span>
                <span class="n">validated_alleles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">validated_allele</span><span class="p">)</span>
                <span class="k">assert</span> <span class="n">validated_allele</span><span class="o">.</span><span class="n">gene</span> <span class="o">==</span> <span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> \
                    <span class="s2">&quot;The allele referring to gene </span><span class="si">{}</span><span class="s2"> is inside gene </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">validated_allele</span><span class="o">.</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">alleles</span> <span class="o">=</span> <span class="n">validated_alleles</span>
        <span class="n">isoforms</span> <span class="o">=</span> <span class="n">mhc2</span><span class="o">.</span><span class="n">isoforms</span>
        <span class="n">validated_isoforms</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">isoform</span> <span class="ow">in</span> <span class="n">isoforms</span><span class="p">:</span>
            <span class="n">validated_isoform</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate_mhc2_isoform_representation</span><span class="p">(</span><span class="n">isoform</span><span class="p">)</span>
            <span class="n">validated_isoforms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">validated_isoform</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">mhc2</span><span class="o">.</span><span class="n">name</span> <span class="o">!=</span> <span class="n">Mhc2Name</span><span class="o">.</span><span class="n">DR</span><span class="p">:</span>
                <span class="k">assert</span> <span class="n">validated_isoform</span><span class="o">.</span><span class="n">alpha_chain</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="p">[</span><span class="n">a</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">alleles</span><span class="p">],</span> \
                    <span class="s2">&quot;Alpha chain allele not present in th list of alleles&quot;</span>
            <span class="k">assert</span> <span class="n">validated_isoform</span><span class="o">.</span><span class="n">beta_chain</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="p">[</span><span class="n">a</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genes</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">alleles</span><span class="p">],</span> \
                <span class="s2">&quot;Beta chain allele not present in th list of alleles&quot;</span>
        <span class="n">mhc2</span><span class="o">.</span><span class="n">isoforms</span> <span class="o">=</span> <span class="n">validated_isoforms</span>
        <span class="k">return</span> <span class="n">mhc2</span>

<div class="viewcode-block" id="ModelValidator.validate_mhc_allele_representation"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.validation.ModelValidator.validate_mhc_allele_representation">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">validate_mhc_allele_representation</span><span class="p">(</span><span class="n">allele</span><span class="p">:</span> <span class="n">MhcAllele</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MhcAllele</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">full_name</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="n">allele</span><span class="o">.</span><span class="n">full_name</span><span class="p">:</span>
                <span class="c1"># infers gene, group and protein from the name</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">HLA_ALLELE_PATTERN</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">allele</span><span class="o">.</span><span class="n">full_name</span><span class="p">)</span>
                <span class="k">assert</span> <span class="n">match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Allele does not match HLA allele pattern </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">allele</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">group</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
                <span class="n">full_name</span> <span class="o">=</span> <span class="n">allele</span><span class="o">.</span><span class="n">full_name</span>
            <span class="k">elif</span> <span class="n">allele</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="c1"># infers gene, group and protein from the name</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">HLA_ALLELE_PATTERN</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">allele</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="k">assert</span> <span class="n">match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Allele does not match HLA allele pattern </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">allele</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">group</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">allele</span><span class="o">.</span><span class="n">gene</span> <span class="ow">and</span> <span class="n">allele</span><span class="o">.</span><span class="n">group</span> <span class="ow">and</span> <span class="n">allele</span><span class="o">.</span><span class="n">protein</span><span class="p">:</span>
                <span class="c1"># infers name from gene, group and protein</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">allele</span><span class="o">.</span><span class="n">gene</span>
                <span class="n">group</span> <span class="o">=</span> <span class="n">allele</span><span class="o">.</span><span class="n">group</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">allele</span><span class="o">.</span><span class="n">protein</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">NeofoxDataValidationException</span><span class="p">(</span>
                    <span class="s2">&quot;HLA allele missing required fields, either name or gene, group and protein must be provided&quot;</span><span class="p">)</span>

            <span class="k">assert</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">Mhc1Name</span><span class="o">.</span><span class="n">__members__</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">Mhc2GeneName</span><span class="o">.</span><span class="n">__members__</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> \
                <span class="s2">&quot;Gene not from classic MHC: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="c1"># builds the final allele representation and validates it just in case</span>
            <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;HLA-</span><span class="si">{gene}</span><span class="s2">*</span><span class="si">{serotype}</span><span class="s2">:</span><span class="si">{protein}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span> <span class="n">serotype</span><span class="o">=</span><span class="n">group</span><span class="p">,</span> <span class="n">protein</span><span class="o">=</span><span class="n">protein</span><span class="p">)</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">HLA_ALLELE_PATTERN</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Allele does not match HLA allele pattern </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">AssertionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">NeofoxDataValidationException</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">MhcAllele</span><span class="p">(</span><span class="n">full_name</span><span class="o">=</span><span class="n">full_name</span> <span class="k">if</span> <span class="n">full_name</span> <span class="k">else</span> <span class="n">name</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">gene</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">group</span><span class="p">,</span> <span class="n">protein</span><span class="o">=</span><span class="n">protein</span><span class="p">)</span></div>

<div class="viewcode-block" id="ModelValidator.validate_mhc2_isoform_representation"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.validation.ModelValidator.validate_mhc2_isoform_representation">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">validate_mhc2_isoform_representation</span><span class="p">(</span><span class="n">isoform</span><span class="p">:</span> <span class="n">Mhc2Isoform</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Mhc2Isoform</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">isoform</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="c1"># infers alpha and beta chains</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">HLA_MOLECULE_PATTERN</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">isoform</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">alpha_chain</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate_mhc_allele_representation</span><span class="p">(</span><span class="n">MhcAllele</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)))</span>
                    <span class="n">beta_chain</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate_mhc_allele_representation</span><span class="p">(</span><span class="n">MhcAllele</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">match</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">HLA_DR_MOLECULE_PATTERN</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">isoform</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                    <span class="k">assert</span> <span class="n">match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Molecule does not match HLA isoform pattern </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">isoform</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                    <span class="n">alpha_chain</span> <span class="o">=</span> <span class="n">MhcAllele</span><span class="p">()</span>
                    <span class="n">beta_chain</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate_mhc_allele_representation</span><span class="p">(</span><span class="n">MhcAllele</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)))</span>
            <span class="k">elif</span> <span class="n">isoform</span><span class="o">.</span><span class="n">alpha_chain</span> <span class="ow">and</span> <span class="n">isoform</span><span class="o">.</span><span class="n">beta_chain</span><span class="p">:</span>
                <span class="c1"># infers name from gene, group and protein</span>
                <span class="n">alpha_chain</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate_mhc_allele_representation</span><span class="p">(</span><span class="n">isoform</span><span class="o">.</span><span class="n">alpha_chain</span><span class="p">)</span>
                <span class="n">beta_chain</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">validate_mhc_allele_representation</span><span class="p">(</span><span class="n">isoform</span><span class="o">.</span><span class="n">beta_chain</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">NeofoxDataValidationException</span><span class="p">(</span><span class="s2">&quot;HLA isoform missing required fields&quot;</span><span class="p">)</span>

            <span class="c1"># builds the final allele representation and validates it just in case</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">get_mhc2_isoform_name</span><span class="p">(</span><span class="n">alpha_chain</span><span class="p">,</span> <span class="n">beta_chain</span><span class="p">)</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">HLA_MOLECULE_PATTERN</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">match2</span> <span class="o">=</span> <span class="n">ModelConverter</span><span class="o">.</span><span class="n">HLA_DR_MOLECULE_PATTERN</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="k">assert</span> <span class="n">match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">match2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> \
                <span class="s2">&quot;Molecule does not match HLA isoform pattern </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">AssertionError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">NeofoxDataValidationException</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">Mhc2Isoform</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">alpha_chain</span><span class="o">=</span><span class="n">alpha_chain</span><span class="p">,</span> <span class="n">beta_chain</span><span class="o">=</span><span class="n">beta_chain</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_validate_expression_values</span><span class="p">(</span><span class="n">neoantigen</span><span class="p">):</span>
        <span class="k">assert</span> <span class="n">neoantigen</span><span class="o">.</span><span class="n">rna_expression</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">neoantigen</span><span class="o">.</span><span class="n">rna_expression</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">,</span> \
            <span class="s2">&quot;RNA expression should be a positive integer or zero </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">neoantigen</span><span class="o">.</span><span class="n">rna_expression</span><span class="p">)</span>
        <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_vaf</span><span class="p">(</span><span class="n">neoantigen</span><span class="o">.</span><span class="n">dna_variant_allele_frequency</span><span class="p">)</span>
        <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_vaf</span><span class="p">(</span><span class="n">neoantigen</span><span class="o">.</span><span class="n">rna_variant_allele_frequency</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_validate_mutation</span><span class="p">(</span><span class="n">mutation</span><span class="p">:</span> <span class="n">Mutation</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Mutation</span><span class="p">:</span>
        <span class="c1"># checks aminoacids</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">mutated_aminoacid</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_aminoacid</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">mutated_aminoacid</span><span class="p">)</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">wild_type_aminoacid</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_aminoacid</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">wild_type_aminoacid</span><span class="p">)</span>

        <span class="c1"># checks left and right flanking regions</span>
        <span class="k">assert</span> <span class="n">mutation</span><span class="o">.</span><span class="n">left_flanking_region</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Empty left flanking region&quot;</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">left_flanking_region</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">left_flanking_region</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">left_flanking_region</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Empty left flanking region&quot;</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">mutation</span><span class="o">.</span><span class="n">left_flanking_region</span><span class="p">:</span>
            <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_aminoacid</span><span class="p">(</span><span class="n">aa</span><span class="p">)</span>

        <span class="k">assert</span> <span class="n">mutation</span><span class="o">.</span><span class="n">right_flanking_region</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Empty right flanking region&quot;</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">right_flanking_region</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">right_flanking_region</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">right_flanking_region</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Empty right flanking region&quot;</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">mutation</span><span class="o">.</span><span class="n">right_flanking_region</span><span class="p">:</span>
            <span class="n">ModelValidator</span><span class="o">.</span><span class="n">_validate_aminoacid</span><span class="p">(</span><span class="n">aa</span><span class="p">)</span>

        <span class="c1"># checks the position</span>
        <span class="k">assert</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Empty position&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="s2">&quot;Position must be an integer&quot;</span>
        <span class="k">assert</span> <span class="n">mutation</span><span class="o">.</span><span class="n">position</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Position must be a 1-based positive integer&quot;</span>
        <span class="k">return</span> <span class="n">mutation</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_validate_transcript</span><span class="p">(</span><span class="n">transcript</span><span class="p">:</span> <span class="n">Transcript</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Transcript</span><span class="p">:</span>

        <span class="c1"># TODO: validate that gene symbol exists</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">gene</span> <span class="k">else</span> <span class="n">transcript</span><span class="o">.</span><span class="n">gene</span>
        <span class="k">assert</span> <span class="n">gene_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Empty gene symbol&quot;</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">gene</span> <span class="o">=</span> <span class="n">gene_name</span>

        <span class="c1"># TODO: validate that transcript identifier exists</span>
        <span class="n">transcript_identifier</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">identifier</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">identifier</span> <span class="k">else</span> <span class="n">transcript</span><span class="o">.</span><span class="n">identifier</span>
        <span class="k">assert</span> <span class="n">transcript_identifier</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">transcript_identifier</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Empty transcript identifier&quot;</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="n">transcript_identifier</span>

        <span class="c1"># TODO: support other assemblies</span>
        <span class="n">assembly</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">assembly</span> <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">assembly</span> <span class="k">else</span> <span class="s2">&quot;hg19&quot;</span>
        <span class="k">assert</span> <span class="n">assembly</span> <span class="o">==</span> <span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;Other reference genome than hg19 is not supported&quot;</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">assembly</span> <span class="o">=</span> <span class="n">assembly</span>

        <span class="k">return</span> <span class="n">transcript</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_validate_vaf</span><span class="p">(</span><span class="n">vaf</span><span class="p">):</span>
        <span class="k">assert</span> <span class="n">vaf</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="mf">0.0</span> <span class="o">&lt;=</span> <span class="n">vaf</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="s2">&quot;VAF should be a positive integer or zero </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">vaf</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_enrich_neoantigen</span><span class="p">(</span><span class="n">neoantigen</span><span class="p">:</span> <span class="n">Neoantigen</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Neoantigen</span><span class="p">:</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">wild_type_xmer</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
            <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">left_flanking_region</span><span class="p">,</span>
            <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">wild_type_aminoacid</span><span class="p">,</span>
            <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">right_flanking_region</span><span class="p">])</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">mutated_xmer</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
            <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">left_flanking_region</span><span class="p">,</span>
            <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">mutated_aminoacid</span><span class="p">,</span>
            <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">right_flanking_region</span><span class="p">])</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">size_left_flanking_region</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">left_flanking_region</span><span class="p">)</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">size_right_flanking_region</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">neoantigen</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">right_flanking_region</span><span class="p">)</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="n">ModelValidator</span><span class="o">.</span><span class="n">generate_neoantigen_identifier</span><span class="p">(</span><span class="n">neoantigen</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">neoantigen</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_validate_aminoacid</span><span class="p">(</span><span class="n">aminoacid</span><span class="p">):</span>
        <span class="k">assert</span> <span class="n">aminoacid</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Aminoacid field cannot be empty&quot;</span>
        <span class="n">aminoacid</span> <span class="o">=</span> <span class="n">aminoacid</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">aminoacid</span><span class="p">,</span> <span class="nb">str</span><span class="p">),</span> <span class="s2">&quot;Aminoacid has to be a string&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">aminoacid</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">aminoacid</span> <span class="ow">in</span> <span class="n">IUPACData</span><span class="o">.</span><span class="n">protein_letters_3to1_extended</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> \
                <span class="s2">&quot;Non existing 3 letter aminoacid </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">aminoacid</span><span class="p">)</span>
            <span class="n">aminoacid</span> <span class="o">=</span> <span class="n">IUPACData</span><span class="o">.</span><span class="n">protein_letters_3to1_extended</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aminoacid</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">aminoacid</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">aminoacid</span> <span class="ow">in</span> <span class="n">ExtendedIUPACProtein</span><span class="o">.</span><span class="n">letters</span><span class="p">,</span> <span class="s2">&quot;Non existing aminoacid </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">aminoacid</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;Invalid aminoacid </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">aminoacid</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">aminoacid</span>

<div class="viewcode-block" id="ModelValidator.generate_neoantigen_identifier"><a class="viewcode-back" href="../../../neofox.model.html#neofox.model.validation.ModelValidator.generate_neoantigen_identifier">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">generate_neoantigen_identifier</span><span class="p">(</span><span class="n">neoantigen</span><span class="p">:</span> <span class="n">Neoantigen</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">neoantigen</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="kc">None</span>    <span class="c1"># this needs to be done otherwise we cannot rreproduce the id after it is set</span>
        <span class="k">return</span> <span class="n">base64</span><span class="o">.</span><span class="n">b64encode</span><span class="p">(</span><span class="n">hashlib</span><span class="o">.</span><span class="n">md5</span><span class="p">(</span><span class="n">neoantigen</span><span class="o">.</span><span class="n">to_json</span><span class="p">()</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s1">&#39;utf8&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">digest</span><span class="p">())</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf8&#39;</span><span class="p">)</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Franziska Lang &amp; Pablo Riesgo Ferreiro

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>