

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>NeoFox - NEOantigen Feature tOolboX &mdash; Neofox 0.4.0.dev1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation instructions" href="install.html" />
    <link rel="prev" title="NeoFox - NEOantigen Feature tOolboX" href="index.html" />
    <link href="_static/style.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Neofox
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">NeoFox - NEOantigen Feature tOolboX</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#neofox-requirements">NeoFox Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#usage-from-the-command-line">Usage from the command line</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#input-data">Input data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output-data">Output data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="models.html">Data models</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_usage.html">Programmatic usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="neofox.html">Neofox package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Neofox</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>NeoFox - NEOantigen Feature tOolboX</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/readme.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="neofox-neoantigen-feature-toolbox">
<h1>NeoFox - NEOantigen Feature tOolboX<a class="headerlink" href="#neofox-neoantigen-feature-toolbox" title="Permalink to this headline">¶</a></h1>
<a class="reference external image-reference" href="https://zenodo.org/badge/latestdoi/294667387"><img alt="DOI" src="https://zenodo.org/badge/294667387.svg" /></a>
<a class="reference external image-reference" href="https://badge.fury.io/py/neofox"><img alt="PyPI version" src="https://badge.fury.io/py/neofox.svg" /></a>
<p>Annotation of mutated peptide sequences (mps) with published neo-epitope descriptors</p>
<p><strong>Published Descriptors:</strong></p>
<ul class="simple">
<li><p>netMHCpan <em>(Jurtz et al, 2017, The Journal of Immunology )</em></p></li>
<li><p>netMHCIIpan <em>(Jensen et al, 2018, Immunology )</em></p></li>
<li><p>IEDB immunogenicity <em>(Calis et al, 2013, PLoS Comput Biol.)</em></p></li>
<li><p>Self-similarity, Conserved vs. Improved Binding  <em>(Bjerregaard et al, 2017, Front Immunol.)</em></p></li>
<li><p>Priority Score <em>(Bjerregaard et al, 2017, Cancer Immunol Immunother.)</em></p></li>
<li><p>DAI <em>(Duan et al., 2014, JEM; Ghorani et al., 2018, Ann Oncol.)</em></p></li>
<li><p>Neoantigen Fitness <em>(Luksza et al., 2017, Nature; Balachandran et al, 2017, Nature)</em></p></li>
<li><p>Residue-centric presentation score (best_rank) &amp; Patient harmonic Best Rank (PHBR-I/II) <em>(Marty et al, 2017, Cell; Marty et al, 2018, Cell)</em></p></li>
<li><p>Classically vs Alternatively Defined Neopitopes &amp; Generator Rate <em>(Rech et al., 2018, Cancer Immunology Research)</em></p></li>
<li><p>Tcell_predictor <em>(Besser et al, 2019, Journal for ImmunoTherapy of Cancer)</em></p></li>
<li><p>neoag <em>(Smith et al, 2019, Cancer Immunology Research)</em></p></li>
<li><p>neoantigen dissimilarity <em>(Richman et al, 2019, Cell Systems)</em></p></li>
<li><p>MixMHCpred <em>(Bassani-Sternberg et al., 2017, PLoS Comp Bio; Gfeller, 2018; J Immunol.)</em></p></li>
<li><p>MixMHC2pred <em>(Racle et al, 2019, Nat. Biotech. 2019)</em></p></li>
<li><p>Vaxrank <em>(Rubinsteyn, 2017, Front Immunol;Wang, 2019, Bioinformatics)</em></p></li>
</ul>
<div class="section" id="neofox-requirements">
<h2>NeoFox Requirements<a class="headerlink" href="#neofox-requirements" title="Permalink to this headline">¶</a></h2>
<p><strong>Required Software/Tools/Dependencies:</strong></p>
<ul class="simple">
<li><p>Python 3.7</p></li>
<li><p>R 3.6.0</p></li>
<li><p>BLAST 2.8.1</p></li>
<li><p>netMHCpan 4.0</p></li>
<li><p>netMHCIIpan 3.2</p></li>
<li><p>MixMHCpred 2.0.2</p></li>
<li><p>MixMHC2pred 1.1.3</p></li>
</ul>
</div>
<div class="section" id="usage-from-the-command-line">
<h2>Usage from the command line<a class="headerlink" href="#usage-from-the-command-line" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">neofox</span> <span class="o">--</span><span class="n">model</span><span class="o">-</span><span class="n">file</span><span class="o">/--</span><span class="n">candidate</span><span class="o">-</span><span class="n">file</span> <span class="n">neoantigens</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">patient</span><span class="o">-</span><span class="nb">id</span> <span class="n">Ptx</span> <span class="o">--</span><span class="n">patient</span><span class="o">-</span><span class="n">data</span> <span class="n">patient_data</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">folder</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">out</span> <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">prefix</span> <span class="n">out_prefix</span> <span class="p">[</span><span class="o">--</span><span class="k">with</span><span class="o">-</span><span class="n">short</span><span class="o">-</span><span class="n">wide</span><span class="o">-</span><span class="n">table</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="k">with</span><span class="o">-</span><span class="n">tall</span><span class="o">-</span><span class="n">skinny</span><span class="o">-</span><span class="n">table</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="k">with</span><span class="o">-</span><span class="n">json</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">num_cpus</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="input-data">
<h3>Input data<a class="headerlink" href="#input-data" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--candidate-file</span></code>: tab-separated values table with neoantigen candidates represented by long mutated peptide sequences</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--model-file</span></code>: tab-separated values table with neoantigens in Neofox model format described in <a class="reference external" href="neofox/model/neoantigen.proto">protobuf model</a></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--patient-id</span></code>: patient identifier (<strong>optional</strong>, this will be used as the patient id for neoantigens without patient)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--patient</span> <span class="pre">data</span></code>: a table of tab separated values containing metadata on the patient</p></li>
</ul>
<p><strong>NOTE</strong>: provide either <code class="docutils literal notranslate"><span class="pre">--candidate-file</span></code> or <code class="docutils literal notranslate"><span class="pre">--model-file</span></code></p>
<p>Example of candidate neoantigens table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span>    <span class="n">UCSC_transcript</span> <span class="n">transcript_expression</span>   <span class="n">substitution</span>    <span class="o">+-</span><span class="mi">13</span><span class="n">_AA_</span><span class="p">(</span><span class="n">SNV</span><span class="p">)</span><span class="n">_</span><span class="o">/</span><span class="n">_</span><span class="o">-</span><span class="mi">15</span><span class="n">_AA_to_STOP_</span><span class="p">(</span><span class="n">INDEL</span><span class="p">)</span>  <span class="p">[</span><span class="n">WT</span><span class="p">]</span><span class="n">_</span><span class="o">+-</span><span class="mi">13</span><span class="n">_AA_</span><span class="p">(</span><span class="n">SNV</span><span class="p">)</span><span class="n">_</span><span class="o">/</span><span class="n">_</span><span class="o">-</span><span class="mi">15</span><span class="n">_AA_to_STOP_</span><span class="p">(</span><span class="n">INDEL</span><span class="p">)</span> <span class="n">VAF_in_tumor</span>    <span class="n">VAF_in_RNA</span>
<span class="n">VCAN</span>    <span class="n">uc003kii</span><span class="o">.</span><span class="mi">3</span>  <span class="mf">0.519506894</span> <span class="n">I547T</span>   <span class="n">DEVLGEPSQDILVTDQTRLEATISPET</span> <span class="n">DEVLGEPSQDILVIDQTRLEATISPET</span> <span class="mf">0.294573643</span> <span class="mf">0.857142857</span>
</pre></div>
</div>
<p>where:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gene</span></code> is the HGNC gene symbol</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">UCSC_trancript</span></code> is the UCSC transcript id including the version</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">substitution</span></code> represents a single aminoacid substitution with single letter aminoacids (eg: I547T)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">+-13_AA_(SNV)_/_-15_AA_to_STOP_(INDEL)</span></code> the mutated neoantigen</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">[WT]_+-13_AA_(SNV)_/_-15_AA_to_STOP_(INDEL)</span></code> the equivalent aminoacid sequence in the normal tissue</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">transcript_expression</span></code> the transcript expression (<strong>optional</strong>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">VAF_in_tumor</span></code> variant allele frequency in the DNA (<strong>optional</strong>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">VAF_in_RNA</span></code> variant allele frequency in the RNA (<strong>optional</strong>, this will be estimated using the <code class="docutils literal notranslate"><span class="pre">VAF_in_tumor</span></code> if not available)</p></li>
</ul>
<p>Example of Neofox model neoantigens table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">transcript</span><span class="o">.</span><span class="n">assembly</span> <span class="n">transcript</span><span class="o">.</span><span class="n">gene</span> <span class="n">transcript</span><span class="o">.</span><span class="n">identifier</span>   <span class="n">mutation</span><span class="o">.</span><span class="n">leftFlankingRegion</span> <span class="n">mutation</span><span class="o">.</span><span class="n">mutatedAminoacid</span>   <span class="n">mutation</span><span class="o">.</span><span class="n">position</span>   <span class="n">mutation</span><span class="o">.</span><span class="n">rightFlankingRegion</span>    <span class="n">mutation</span><span class="o">.</span><span class="n">wildTypeAminoacid</span>  <span class="n">patientIdentifier</span>   <span class="n">rnaExpression</span>   <span class="n">rnaVariantAlleleFrequency</span>   <span class="n">dnaVariantAlleleFrequency</span>
<span class="n">hg19</span>    <span class="n">BRCA2</span>   <span class="n">uc003kii</span><span class="o">.</span><span class="mi">3</span>  <span class="n">AAAAAA</span>  <span class="n">L</span>   <span class="mi">935</span> <span class="n">AAAAA</span>   <span class="n">F</span>   <span class="n">Pt1</span> <span class="mf">4.512</span>   <span class="mf">0.4675</span>  <span class="mf">0.36103</span>
<span class="n">hg19</span>    <span class="n">BRCA2</span>   <span class="n">uc003kii</span><span class="o">.</span><span class="mi">3</span>  <span class="n">AAAAAA</span>  <span class="n">M</span>   <span class="mi">518</span> <span class="n">AAAAA</span>   <span class="n">R</span>   <span class="n">Pt2</span> <span class="mf">0.154</span>   <span class="mf">0.015404</span>    <span class="mf">0.034404</span>
<span class="n">hg19</span>    <span class="n">BRCA2</span>   <span class="n">uc003kii</span><span class="o">.</span><span class="mi">3</span>  <span class="n">AAAAAA</span>  <span class="n">G</span>   <span class="mi">285</span> <span class="n">AAAAA</span>   <span class="n">K</span>   <span class="n">Pt3</span> <span class="mf">8.841207</span>    <span class="mf">0.89387</span> <span class="mf">0.51924</span>
</pre></div>
</div>
<p>where:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">transcript.assembly</span></code> - the assembly of the reference genome (only hg19 is supported)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">transcript.gene</span></code> - the HGMC gene symbol</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">transcript.identifier</span></code> - the UCSC transcript identifier including the version number</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mutation.leftFlankingRegion</span></code> - the aminoacids flanking the mutation on the left (in IUPAC one letter symbols)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mutation.mutatedAminoacid</span></code> - the mutated aminoacid (IUPAC 1 or 3 letters respecting casing, eg: A and Ala)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mutation.position</span></code> - the 1 based position of the mutation in the protein</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mutation.rightFlankingRegion</span></code> - the aminoacids flanking the mutation on the left (in IUPAC one letter symbols)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mutation.wildTypeAminoacid</span></code> - the wild type aminoacid (IUPAC 1 or 3 letters respecting casing, eg: A and Ala)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">patientIdentifier</span></code> - the patient identifier</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rnaExpression</span></code> - the transcript RNA expression (<strong>optional</strong>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rnaVariantAlleleFrequency</span></code> - the variant allele frequency calculated from the RNA (<strong>optional</strong>, this will be estimated using the <code class="docutils literal notranslate"><span class="pre">dnaVariantAlleleFrequency</span></code> if not available)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">dnaVariantAlleleFrequency</span></code> - the variant allele frequency calculated from the DNA (<strong>optional</strong>)</p></li>
</ul>
<p>Example of patients data table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">identifier</span>  <span class="n">mhcIAlleles</span> <span class="n">mhcIIAlleles</span>   <span class="n">isRnaAvailable</span>
<span class="n">Pt29</span>    <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span><span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span><span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">07</span><span class="p">:</span><span class="mi">02</span> <span class="n">HLA</span><span class="o">-</span><span class="n">DRB1</span><span class="o">*</span><span class="mi">11</span><span class="p">:</span><span class="mi">04</span><span class="p">,</span><span class="n">HLA</span><span class="o">-</span><span class="n">DRB1</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">01</span>  <span class="kc">True</span>
</pre></div>
</div>
<p>where:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">identifier</span></code> - the patient identifier</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mhcIAlleles</span></code> - the list of MHC I alleles in the patient</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mhcIIAlleles</span></code> - the list of MHC II alleles in the patient</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">isRnaAvailable</span></code> - whether RNA was available for the analysis. If true then the <code class="docutils literal notranslate"><span class="pre">VAF_in_RNA</span></code> field will be used, else <code class="docutils literal notranslate"><span class="pre">VAF_in_DNA</span></code> will be used. (<strong>optional</strong>)</p></li>
</ul>
</div>
<div class="section" id="output-data">
<h3>Output data<a class="headerlink" href="#output-data" title="Permalink to this headline">¶</a></h3>
<p>The output data is returned in a short wide tab separated values file (<code class="docutils literal notranslate"><span class="pre">--with-short-wide-table</span></code>). Optionally, it can be provided in a tall skinny tab separated values file (<code class="docutils literal notranslate"><span class="pre">--with-tall-skinny-table</span></code>) or in JSON (<code class="docutils literal notranslate"><span class="pre">--with-json</span></code>).</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="install.html" class="btn btn-neutral float-right" title="Installation instructions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="NeoFox - NEOantigen Feature tOolboX" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Franziska Lang &amp; Pablo Riesgo Ferreiro

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>