{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Programmatic usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "from neofox.model.conversion import ModelConverter\n",
    "from neofox.model.validation import ModelValidator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a neoantigen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from neofox.model.neoantigen import Neoantigen, Transcript, Mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript = Transcript(assembly=\"hg19\", gene=\"VCAN\", identifier=\"uc003kii.3\")\n",
    "mutation = Mutation(position=1007, wild_type_aminoacid=\"I\", mutated_aminoacid=\"T\", left_flanking_region=\"DEVLGEPSQDILV\", right_flanking_region=\"DQTRLEATISPET\")\n",
    "neoantigen = Neoantigen(transcript=transcript, mutation=mutation, patient_identifier=\"P123\", rna_expression=0.519506894, rna_variant_allele_frequency=0.857142857, dna_variant_allele_frequency=0.294573643)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Validate a neoantigen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"identifier\": \"jETwpX0R9iEiQz2SMpHkPQ==\",\n",
      "  \"patientIdentifier\": \"P123\",\n",
      "  \"transcript\": {\n",
      "    \"identifier\": \"uc003kii.3\",\n",
      "    \"assembly\": \"hg19\",\n",
      "    \"gene\": \"VCAN\"\n",
      "  },\n",
      "  \"mutation\": {\n",
      "    \"position\": 1007,\n",
      "    \"wildTypeXmer\": \"DEVLGEPSQDILVIDQTRLEATISPET\",\n",
      "    \"wildTypeAminoacid\": \"I\",\n",
      "    \"mutatedXmer\": \"DEVLGEPSQDILVTDQTRLEATISPET\",\n",
      "    \"mutatedAminoacid\": \"T\",\n",
      "    \"leftFlankingRegion\": \"DEVLGEPSQDILV\",\n",
      "    \"sizeLeftFlankingRegion\": 13,\n",
      "    \"rightFlankingRegion\": \"DQTRLEATISPET\",\n",
      "    \"sizeRightFlankingRegion\": 13\n",
      "  },\n",
      "  \"rnaExpression\": 0.519506894,\n",
      "  \"dnaVariantAlleleFrequency\": 0.294573643,\n",
      "  \"rnaVariantAlleleFrequency\": 0.857142857\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "validated_neoantigen = ModelValidator.validate_neoantigen(neoantigen=neoantigen)\n",
    "print(validated_neoantigen.to_json(indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from neofox.model.neoantigen import Patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the MHC I alleles of the patient and parse them into the object representation\n",
    "mhc1 = ModelConverter.parse_mhc1_alleles([\"HLA-A*01:01:02:03N\", \"HLA-A*01:02:02:03N\", \"HLA-B*01:01:02:03N\", \"HLA-B*01:01:02:04N\", \"HLA-C*01:01\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"zygosity\": \"HETEROZYGOUS\",\n",
      "  \"alleles\": [\n",
      "    {\n",
      "      \"fullName\": \"HLA-A*01:01:02:03N\",\n",
      "      \"name\": \"HLA-A*01:01\",\n",
      "      \"gene\": \"A\",\n",
      "      \"group\": \"01\",\n",
      "      \"protein\": \"01\"\n",
      "    },\n",
      "    {\n",
      "      \"fullName\": \"HLA-A*01:02:02:03N\",\n",
      "      \"name\": \"HLA-A*01:02\",\n",
      "      \"gene\": \"A\",\n",
      "      \"group\": \"01\",\n",
      "      \"protein\": \"02\"\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# MHC allele names will be normalised into 4 digits names\n",
    "print(mhc1[0].to_json(indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the MHC II alleles of the patient and parse them into the object representation\n",
    "mhc2 = ModelConverter.parse_mhc2_alleles([\"HLA-DPA1*01:01\", \"HLA-DPA1*01:02\", \"HLA-DPB1*01:01\", \"HLA-DPB1*01:01\", \"HLA-DRB1*01:01\", \"HLA-DRB1*01:01\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute '_serialized_on_wire'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-ab8478a95705>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmhc2\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_json\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindent\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/src/addannot/venv/lib/python3.6/site-packages/betterproto/__init__.py\u001b[0m in \u001b[0;36mto_json\u001b[0;34m(self, indent)\u001b[0m\n\u001b[1;32m    925\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mto_json\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindent\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mUnion\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    926\u001b[0m         \u001b[0;34m\"\"\"Returns the encoded JSON representation of this message instance.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 927\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdumps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindent\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindent\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    928\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    929\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mfrom_json\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mT\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mUnion\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbytes\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mT\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/src/addannot/venv/lib/python3.6/site-packages/betterproto/__init__.py\u001b[0m in \u001b[0;36mto_dict\u001b[0;34m(self, casing, include_default_values)\u001b[0m\n\u001b[1;32m    827\u001b[0m                 \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    828\u001b[0m                     \u001b[0;31m# Convert each item.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 829\u001b[0;31m                     \u001b[0mv\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasing\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minclude_default_values\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    830\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0minclude_default_values\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    831\u001b[0m                         \u001b[0moutput\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcased_name\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/src/addannot/venv/lib/python3.6/site-packages/betterproto/__init__.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    827\u001b[0m                 \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    828\u001b[0m                     \u001b[0;31m# Convert each item.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 829\u001b[0;31m                     \u001b[0mv\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasing\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minclude_default_values\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    830\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0minclude_default_values\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    831\u001b[0m                         \u001b[0moutput\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcased_name\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/src/addannot/venv/lib/python3.6/site-packages/betterproto/__init__.py\u001b[0m in \u001b[0;36mto_dict\u001b[0;34m(self, casing, include_default_values)\u001b[0m\n\u001b[1;32m    831\u001b[0m                         \u001b[0moutput\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcased_name\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    832\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 833\u001b[0;31m                     \u001b[0;32mif\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_serialized_on_wire\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0minclude_default_values\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    834\u001b[0m                         \u001b[0moutput\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcased_name\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasing\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minclude_default_values\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    835\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mmeta\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mproto_type\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"map\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute '_serialized_on_wire'"
     ]
    }
   ],
   "source": [
    "print(mhc2[0].to_json(indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mhc2(name=<Mhc2Name.DR: 0>, genes=[Mhc2Gene(name=<Mhc2GeneName.DRB1: 0>, zygosity=<Zygosity.HOMOZYGOUS: 0>, alleles=[MhcAllele(full_name='HLA-DRB1*01:01', name='HLA-DRB1*01:01', gene='DRB1', group='01', protein='01')])], isoforms=[Mhc2Isoform(name='HLA-DRB1*01:01', alpha_chain=None, beta_chain=MhcAllele(full_name='HLA-DRB1*01:01', name='HLA-DRB1*01:01', gene='DRB1', group='01', protein='01'))])"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mhc2[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "patient = Patient(identifier=\"P123\", is_rna_available=True, mhc1=mhc1, mhc2=mhc2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Validate a patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from neofox.model.validation import ModelValidator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "validated_patient = ModelValidator.validate_patient(patient)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Neofox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject\n",
      "  return f(*args, **kwds)\n",
      "/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject\n",
      "  return f(*args, **kwds)\n",
      "/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject\n",
      "  return f(*args, **kwds)\n"
     ]
    }
   ],
   "source": [
    "from neofox.neofox import NeoFox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[I 201020 22:58:59 neofox:80] Loading data...\n",
      "/home/priesgo/src/addannot/venv/lib/python3.6/site-packages/distributed/node.py:173: UserWarning: Port 8787 is already in use.\n",
      "Perhaps you already have a cluster running?\n",
      "Hosting the HTTP server on port 46855 instead\n",
      "  http_address[\"port\"], self.http_server.port\n",
      "[I 201020 22:59:00 references:141] Reference genome folder: /home/priesgo/neofox_install/reference_data_5\n",
      "[I 201020 22:59:00 references:142] Resources\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/netmhc2pan_available_alleles.txt\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/netmhcpan_available_alleles.txt\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/iedb\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/proteome_db\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/proteome_db/Homo_sapiens.fa\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/iedb/IEDB.fasta\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/iedb/iedb_blast_db.phr\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/iedb/iedb_blast_db.pin\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/iedb/iedb_blast_db.psq\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/proteome_db/Homo_sapiens.fa\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/proteome_db/homo_sapiens.phr\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/proteome_db/homo_sapiens.pin\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/proteome_db/homo_sapiens.pog\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/proteome_db/homo_sapiens.psd\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/proteome_db/homo_sapiens.psi\n",
      "[I 201020 22:59:00 references:144] /home/priesgo/neofox_install/reference_data_5/proteome_db/homo_sapiens.psq\n",
      "[I 201020 22:59:00 neofox:69] Data loaded\n",
      "[I 201020 22:59:00 neofox:109] Starting NeoFox annotations...\n",
      "[D 201020 22:59:00 neofox:115] Neoantigen: {\n",
      "       \"identifier\": \"jETwpX0R9iEiQz2SMpHkPQ==\",\n",
      "       \"patientIdentifier\": \"P123\",\n",
      "       \"transcript\": {\n",
      "          \"identifier\": \"uc003kii.3\",\n",
      "          \"assembly\": \"hg19\",\n",
      "          \"gene\": \"VCAN\"\n",
      "       },\n",
      "       \"mutation\": {\n",
      "          \"position\": 1007,\n",
      "          \"wildTypeXmer\": \"DEVLGEPSQDILVIDQTRLEATISPET\",\n",
      "          \"wildTypeAminoacid\": \"I\",\n",
      "          \"mutatedXmer\": \"DEVLGEPSQDILVTDQTRLEATISPET\",\n",
      "          \"mutatedAminoacid\": \"T\",\n",
      "          \"leftFlankingRegion\": \"DEVLGEPSQDILV\",\n",
      "          \"sizeLeftFlankingRegion\": 13,\n",
      "          \"rightFlankingRegion\": \"DQTRLEATISPET\",\n",
      "          \"sizeRightFlankingRegion\": 13\n",
      "       },\n",
      "       \"rnaExpression\": 0.519506894,\n",
      "       \"dnaVariantAlleleFrequency\": 0.294573643,\n",
      "       \"rnaVariantAlleleFrequency\": 0.857142857\n",
      "    }\n",
      "[D 201020 22:59:00 neofox:116] Patient: {\n",
      "       \"identifier\": \"P123\",\n",
      "       \"isRnaAvailable\": true,\n",
      "       \"mhc1\": [\n",
      "          {\n",
      "             \"zygosity\": \"HETEROZYGOUS\",\n",
      "             \"alleles\": [\n",
      "                {\n",
      "                   \"fullName\": \"HLA-A*01:01:02:03N\",\n",
      "                   \"name\": \"HLA-A*01:01\",\n",
      "                   \"gene\": \"A\",\n",
      "                   \"group\": \"01\",\n",
      "                   \"protein\": \"01\"\n",
      "                },\n",
      "                {\n",
      "                   \"fullName\": \"HLA-A*01:02:02:03N\",\n",
      "                   \"name\": \"HLA-A*01:02\",\n",
      "                   \"gene\": \"A\",\n",
      "                   \"group\": \"01\",\n",
      "                   \"protein\": \"02\"\n",
      "                }\n",
      "             ]\n",
      "          },\n",
      "          {\n",
      "             \"name\": \"B\",\n",
      "             \"alleles\": [\n",
      "                {\n",
      "                   \"fullName\": \"HLA-B*01:01:02:04N\",\n",
      "                   \"name\": \"HLA-B*01:01\",\n",
      "                   \"gene\": \"B\",\n",
      "                   \"group\": \"01\",\n",
      "                   \"protein\": \"01\"\n",
      "                }\n",
      "             ]\n",
      "          },\n",
      "          {\n",
      "             \"name\": \"C\",\n",
      "             \"zygosity\": \"HEMIZYGOUS\",\n",
      "             \"alleles\": [\n",
      "                {\n",
      "                   \"fullName\": \"HLA-C*01:01\",\n",
      "                   \"name\": \"HLA-C*01:01\",\n",
      "                   \"gene\": \"C\",\n",
      "                   \"group\": \"01\",\n",
      "                   \"protein\": \"01\"\n",
      "                }\n",
      "             ]\n",
      "          }\n",
      "       ],\n",
      "       \"mhc2\": [\n",
      "          {\n",
      "             \"genes\": [\n",
      "                {\n",
      "                   \"alleles\": [\n",
      "                      {\n",
      "                         \"fullName\": \"HLA-DRB1*01:01\",\n",
      "                         \"name\": \"HLA-DRB1*01:01\",\n",
      "                         \"gene\": \"DRB1\",\n",
      "                         \"group\": \"01\",\n",
      "                         \"protein\": \"01\"\n",
      "                      }\n",
      "                   ]\n",
      "                }\n",
      "             ],\n",
      "             \"isoforms\": [\n",
      "                {\n",
      "                   \"name\": \"HLA-DRB1*01:01\",\n",
      "                   \"betaChain\": {\n",
      "                      \"fullName\": \"HLA-DRB1*01:01\",\n",
      "                      \"name\": \"HLA-DRB1*01:01\",\n",
      "                      \"gene\": \"DRB1\",\n",
      "                      \"group\": \"01\",\n",
      "                      \"protein\": \"01\"\n",
      "                   }\n",
      "                }\n",
      "             ]\n",
      "          },\n",
      "          {\n",
      "             \"name\": \"DP\",\n",
      "             \"genes\": [\n",
      "                {\n",
      "                   \"name\": \"DPA1\",\n",
      "                   \"zygosity\": \"HETEROZYGOUS\",\n",
      "                   \"alleles\": [\n",
      "                      {\n",
      "                         \"fullName\": \"HLA-DPA1*01:01\",\n",
      "                         \"name\": \"HLA-DPA1*01:01\",\n",
      "                         \"gene\": \"DPA1\",\n",
      "                         \"group\": \"01\",\n",
      "                         \"protein\": \"01\"\n",
      "                      },\n",
      "                      {\n",
      "                         \"fullName\": \"HLA-DPA1*01:02\",\n",
      "                         \"name\": \"HLA-DPA1*01:02\",\n",
      "                         \"gene\": \"DPA1\",\n",
      "                         \"group\": \"01\",\n",
      "                         \"protein\": \"02\"\n",
      "                      }\n",
      "                   ]\n",
      "                },\n",
      "                {\n",
      "                   \"name\": \"DPB1\",\n",
      "                   \"alleles\": [\n",
      "                      {\n",
      "                         \"fullName\": \"HLA-DPB1*01:01\",\n",
      "                         \"name\": \"HLA-DPB1*01:01\",\n",
      "                         \"gene\": \"DPB1\",\n",
      "                         \"group\": \"01\",\n",
      "                         \"protein\": \"01\"\n",
      "                      }\n",
      "                   ]\n",
      "                }\n",
      "             ],\n",
      "             \"isoforms\": [\n",
      "                {\n",
      "                   \"name\": \"HLA-DPA1*01:01-DPB1*01:01\",\n",
      "                   \"alphaChain\": {\n",
      "                      \"fullName\": \"HLA-DPA1*01:01\",\n",
      "                      \"name\": \"HLA-DPA1*01:01\",\n",
      "                      \"gene\": \"DPA1\",\n",
      "                      \"group\": \"01\",\n",
      "                      \"protein\": \"01\"\n",
      "                   },\n",
      "                   \"betaChain\": {\n",
      "                      \"fullName\": \"HLA-DPB1*01:01\",\n",
      "                      \"name\": \"HLA-DPB1*01:01\",\n",
      "                      \"gene\": \"DPB1\",\n",
      "                      \"group\": \"01\",\n",
      "                      \"protein\": \"01\"\n",
      "                   }\n",
      "                },\n",
      "                {\n",
      "                   \"name\": \"HLA-DPA1*01:02-DPB1*01:01\",\n",
      "                   \"alphaChain\": {\n",
      "                      \"fullName\": \"HLA-DPA1*01:02\",\n",
      "                      \"name\": \"HLA-DPA1*01:02\",\n",
      "                      \"gene\": \"DPA1\",\n",
      "                      \"group\": \"01\",\n",
      "                      \"protein\": \"02\"\n",
      "                   },\n",
      "                   \"betaChain\": {\n",
      "                      \"fullName\": \"HLA-DPB1*01:01\",\n",
      "                      \"name\": \"HLA-DPB1*01:01\",\n",
      "                      \"gene\": \"DPB1\",\n",
      "                      \"group\": \"01\",\n",
      "                      \"protein\": \"01\"\n",
      "                   }\n",
      "                }\n",
      "             ]\n",
      "          },\n",
      "          {\n",
      "             \"name\": \"DQ\",\n",
      "             \"genes\": [\n",
      "                {\n",
      "                   \"name\": \"DQA1\",\n",
      "                   \"zygosity\": \"LOSS\"\n",
      "                },\n",
      "                {\n",
      "                   \"name\": \"DQB1\",\n",
      "                   \"zygosity\": \"LOSS\"\n",
      "                }\n",
      "             ]\n",
      "          }\n",
      "       ]\n",
      "    }\n",
      "[I 201020 22:59:07 neofox:123] Elapsed time for annotating 1 neoantigens 7 seconds\n"
     ]
    }
   ],
   "source": [
    "annotations = NeoFox(neoantigens=[validated_neoantigen], patients=[validated_patient], patient_id=\"P123\", num_cpus=2).get_annotations()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[NeoantigenAnnotations(neoantigen_identifier='jETwpX0R9iEiQz2SMpHkPQ==', annotations=[Annotation(name='Expression_mutated_transcript', value='0.44529'), Annotation(name='mutation_not_found_in_proteome', value='1'), Annotation(name='Best_rank_MHCI_score', value='0.4384'), Annotation(name='Best_rank_MHCI_score_epitope', value='VTDQTRLEA'), Annotation(name='Best_rank_MHCI_score_allele', value='HLA-A*01:01'), Annotation(name='Best_affinity_MHCI_score', value='874.8'), Annotation(name='Best_affinity_MHCI_epitope', value='VTDQTRLEA'), Annotation(name='Best_affinity_MHCI_allele', value='HLA-A*01:01'), Annotation(name='Best_rank_MHCI_9mer_score', value='0.4384'), Annotation(name='Best_rank_MHCI_9mer_epitope', value='VTDQTRLEA'), Annotation(name='Best_rank_MHCI_9mer_allele', value='HLA-A*01:01'), Annotation(name='Best_affinity_MHCI_9mer_score', value='874.8'), Annotation(name='Best_affinity_MHCI_9mer_allele', value='HLA-A*01:01'), Annotation(name='Best_affinity_MHCI_9mer_epitope', value='VTDQTRLEA'), Annotation(name='Best_affinity_MHCI_score_WT', value='6296.8'), Annotation(name='Best_affinity_MHCI_epitope_WT', value='VIDQTRLEA'), Annotation(name='Best_affinity_MHCI_allele_WT', value='HLA-A*01:01'), Annotation(name='Best_rank_MHCI_score_WT', value='1.8897'), Annotation(name='Best_rank_MHCI_score_epitope_WT', value='VIDQTRLEA'), Annotation(name='Best_rank_MHCI_score_allele_WT', value='HLA-A*01:01'), Annotation(name='Best_rank_MHCI_9mer_score_WT', value='1.8897'), Annotation(name='Best_rank_MHCI_9mer_epitope_WT', value='VIDQTRLEA'), Annotation(name='Best_rank_MHCI_9mer_allele_WT', value='HLA-A*01:01'), Annotation(name='Best_affinity_MHCI_9mer_score_WT', value='6296.8'), Annotation(name='Best_affinity_MHCI_9mer_allele_WT', value='HLA-A*01:01'), Annotation(name='Best_affinity_MHCI_9mer_epitope_WT', value='VIDQTRLEA'), Annotation(name='Generator_rate', value='0'), Annotation(name='PHBR-I', value='NA'), Annotation(name='Best_affinity_MHCI_9mer_position_mutation', value='2'), Annotation(name='Best_rank_MHCII_score', value='65'), Annotation(name='Best_rank_MHCII_score_epitope', value='SQDILVTDQTRLEAT'), Annotation(name='Best_rank_MHCII_score_allele', value='DRB1_0101'), Annotation(name='Best_affinity_MHCII_score', value='647.81'), Annotation(name='Best_affinity_MHCII_epitope', value='SQDILVTDQTRLEAT'), Annotation(name='Best_affinity_MHCII_allele', value='DRB1_0101'), Annotation(name='Best_rank_MHCII_score_WT', value='46'), Annotation(name='Best_rank_MHCII_score_epitope_WT', value='SQDILVIDQTRLEAT'), Annotation(name='Best_rank_MHCII_score_allele_WT', value='DRB1_0101'), Annotation(name='Best_affinity_MHCII_score_WT', value='269.19'), Annotation(name='Best_affinity_MHCII_epitope_WT', value='SQDILVIDQTRLEAT'), Annotation(name='Best_affinity_MHCII_allele_WT', value='DRB1_0101'), Annotation(name='PHBR-II', value='NA'), Annotation(name='Amplitude_MHCI_affinity_9mer', value='2.4915'), Annotation(name='Amplitude_MHCI_affinity', value='2.4915'), Annotation(name='Pathogensimiliarity_MHCI_affinity_9mer', value='0'), Annotation(name='Recognition_Potential_MHCI_affinity_9mer', value='NA'), Annotation(name='DAI_MHCI_affinity_cutoff500nM', value='NA'), Annotation(name='CDN_MHCI', value='0'), Annotation(name='ADN_MHCI', value='0'), Annotation(name='CDN_MHCII', value='0'), Annotation(name='ADN_MHCII', value='0'), Annotation(name='Tcell_predictor_score_cutoff500nM', value='NA'), Annotation(name='Improved_Binder_MHCI', value='1'), Annotation(name='Selfsimilarity_MHCI_conserved_binder', value='NA'), Annotation(name='Number_of_mismatches_MCHI', value='1'), Annotation(name='Priority_score', value='0.09596'), Annotation(name='Neoag_immunogenicity', value='NA'), Annotation(name='IEDB_Immunogenicity_MHCI_cutoff500nM', value='NA'), Annotation(name='MixMHCpred_best_peptide', value='NA'), Annotation(name='MixMHCpred_best_score', value='NA'), Annotation(name='MixMHCpred_best_rank', value='NA'), Annotation(name='MixMHCpred_best_allele', value='NA'), Annotation(name='MixMHC2pred_best_peptide', value='DEVLGEPSQDILVT'), Annotation(name='MixMHC2pred_best_rank', value='2.818'), Annotation(name='MixMHC2pred_best_allele', value='DRB1_01_01'), Annotation(name='Dissimilarity_MHCI_cutoff500nM', value='NA'), Annotation(name='vaxrank_binding_score', value='0.0322'), Annotation(name='vaxrank_total_score', value='0.01434')], annotator='Neofox', annotator_version='0.4.0.dev1', timestamp='20201020225902062657', resources_hash='')]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transform the annotations into a data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations_ts = ModelConverter.annotations2tall_skinny_table(neoantigen_annotations=annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>value</th>\n",
       "      <th>neoantigen_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Expression_mutated_transcript</td>\n",
       "      <td>0.44529</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mutation_not_found_in_proteome</td>\n",
       "      <td>1</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Best_rank_MHCI_score</td>\n",
       "      <td>0.4384</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Best_rank_MHCI_score_epitope</td>\n",
       "      <td>VTDQTRLEA</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Best_rank_MHCI_score_allele</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Best_affinity_MHCI_score</td>\n",
       "      <td>874.8</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Best_affinity_MHCI_epitope</td>\n",
       "      <td>VTDQTRLEA</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Best_affinity_MHCI_allele</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Best_rank_MHCI_9mer_score</td>\n",
       "      <td>0.4384</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Best_rank_MHCI_9mer_epitope</td>\n",
       "      <td>VTDQTRLEA</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Best_rank_MHCI_9mer_allele</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Best_affinity_MHCI_9mer_score</td>\n",
       "      <td>874.8</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Best_affinity_MHCI_9mer_allele</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Best_affinity_MHCI_9mer_epitope</td>\n",
       "      <td>VTDQTRLEA</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Best_affinity_MHCI_score_WT</td>\n",
       "      <td>6296.8</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Best_affinity_MHCI_epitope_WT</td>\n",
       "      <td>VIDQTRLEA</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Best_affinity_MHCI_allele_WT</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Best_rank_MHCI_score_WT</td>\n",
       "      <td>1.8897</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Best_rank_MHCI_score_epitope_WT</td>\n",
       "      <td>VIDQTRLEA</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Best_rank_MHCI_score_allele_WT</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Best_rank_MHCI_9mer_score_WT</td>\n",
       "      <td>1.8897</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Best_rank_MHCI_9mer_epitope_WT</td>\n",
       "      <td>VIDQTRLEA</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Best_rank_MHCI_9mer_allele_WT</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Best_affinity_MHCI_9mer_score_WT</td>\n",
       "      <td>6296.8</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Best_affinity_MHCI_9mer_allele_WT</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Best_affinity_MHCI_9mer_epitope_WT</td>\n",
       "      <td>VIDQTRLEA</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Generator_rate</td>\n",
       "      <td>0</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Best_affinity_MHCI_9mer_position_mutation</td>\n",
       "      <td>2</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Best_rank_MHCII_score</td>\n",
       "      <td>65</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Best_rank_MHCII_score_epitope</td>\n",
       "      <td>SQDILVTDQTRLEAT</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Best_rank_MHCII_score_allele</td>\n",
       "      <td>DRB1_0101</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Best_affinity_MHCII_score</td>\n",
       "      <td>647.81</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Best_affinity_MHCII_epitope</td>\n",
       "      <td>SQDILVTDQTRLEAT</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Best_affinity_MHCII_allele</td>\n",
       "      <td>DRB1_0101</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Best_rank_MHCII_score_WT</td>\n",
       "      <td>46</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Best_rank_MHCII_score_epitope_WT</td>\n",
       "      <td>SQDILVIDQTRLEAT</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Best_rank_MHCII_score_allele_WT</td>\n",
       "      <td>DRB1_0101</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Best_affinity_MHCII_score_WT</td>\n",
       "      <td>269.19</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Best_affinity_MHCII_epitope_WT</td>\n",
       "      <td>SQDILVIDQTRLEAT</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Best_affinity_MHCII_allele_WT</td>\n",
       "      <td>DRB1_0101</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Amplitude_MHCI_affinity_9mer</td>\n",
       "      <td>2.4915</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Amplitude_MHCI_affinity</td>\n",
       "      <td>2.4915</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Pathogensimiliarity_MHCI_affinity_9mer</td>\n",
       "      <td>0</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CDN_MHCI</td>\n",
       "      <td>0</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>ADN_MHCI</td>\n",
       "      <td>0</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CDN_MHCII</td>\n",
       "      <td>0</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>ADN_MHCII</td>\n",
       "      <td>0</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Improved_Binder_MHCI</td>\n",
       "      <td>1</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Number_of_mismatches_MCHI</td>\n",
       "      <td>1</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Priority_score</td>\n",
       "      <td>0.09596</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>MixMHC2pred_best_peptide</td>\n",
       "      <td>DEVLGEPSQDILVT</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>MixMHC2pred_best_rank</td>\n",
       "      <td>2.818</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>MixMHC2pred_best_allele</td>\n",
       "      <td>DRB1_01_01</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>vaxrank_binding_score</td>\n",
       "      <td>0.0322</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>vaxrank_total_score</td>\n",
       "      <td>0.01434</td>\n",
       "      <td>jETwpX0R9iEiQz2SMpHkPQ==</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         name            value     neoantigen_identifier\n",
       "0               Expression_mutated_transcript          0.44529  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "1              mutation_not_found_in_proteome                1  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "2                        Best_rank_MHCI_score           0.4384  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "3                Best_rank_MHCI_score_epitope        VTDQTRLEA  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "4                 Best_rank_MHCI_score_allele      HLA-A*01:01  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "5                    Best_affinity_MHCI_score            874.8  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "6                  Best_affinity_MHCI_epitope        VTDQTRLEA  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "7                   Best_affinity_MHCI_allele      HLA-A*01:01  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "8                   Best_rank_MHCI_9mer_score           0.4384  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "9                 Best_rank_MHCI_9mer_epitope        VTDQTRLEA  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "10                 Best_rank_MHCI_9mer_allele      HLA-A*01:01  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "11              Best_affinity_MHCI_9mer_score            874.8  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "12             Best_affinity_MHCI_9mer_allele      HLA-A*01:01  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "13            Best_affinity_MHCI_9mer_epitope        VTDQTRLEA  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "14                Best_affinity_MHCI_score_WT           6296.8  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "15              Best_affinity_MHCI_epitope_WT        VIDQTRLEA  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "16               Best_affinity_MHCI_allele_WT      HLA-A*01:01  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "17                    Best_rank_MHCI_score_WT           1.8897  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "18            Best_rank_MHCI_score_epitope_WT        VIDQTRLEA  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "19             Best_rank_MHCI_score_allele_WT      HLA-A*01:01  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "20               Best_rank_MHCI_9mer_score_WT           1.8897  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "21             Best_rank_MHCI_9mer_epitope_WT        VIDQTRLEA  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "22              Best_rank_MHCI_9mer_allele_WT      HLA-A*01:01  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "23           Best_affinity_MHCI_9mer_score_WT           6296.8  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "24          Best_affinity_MHCI_9mer_allele_WT      HLA-A*01:01  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "25         Best_affinity_MHCI_9mer_epitope_WT        VIDQTRLEA  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "26                             Generator_rate                0  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "28  Best_affinity_MHCI_9mer_position_mutation                2  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "29                      Best_rank_MHCII_score               65  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "30              Best_rank_MHCII_score_epitope  SQDILVTDQTRLEAT  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "31               Best_rank_MHCII_score_allele        DRB1_0101  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "32                  Best_affinity_MHCII_score           647.81  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "33                Best_affinity_MHCII_epitope  SQDILVTDQTRLEAT  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "34                 Best_affinity_MHCII_allele        DRB1_0101  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "35                   Best_rank_MHCII_score_WT               46  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "36           Best_rank_MHCII_score_epitope_WT  SQDILVIDQTRLEAT  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "37            Best_rank_MHCII_score_allele_WT        DRB1_0101  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "38               Best_affinity_MHCII_score_WT           269.19  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "39             Best_affinity_MHCII_epitope_WT  SQDILVIDQTRLEAT  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "40              Best_affinity_MHCII_allele_WT        DRB1_0101  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "42               Amplitude_MHCI_affinity_9mer           2.4915  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "43                    Amplitude_MHCI_affinity           2.4915  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "44     Pathogensimiliarity_MHCI_affinity_9mer                0  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "47                                   CDN_MHCI                0  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "48                                   ADN_MHCI                0  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "49                                  CDN_MHCII                0  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "50                                  ADN_MHCII                0  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "52                       Improved_Binder_MHCI                1  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "54                  Number_of_mismatches_MCHI                1  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "55                             Priority_score          0.09596  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "62                   MixMHC2pred_best_peptide   DEVLGEPSQDILVT  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "63                      MixMHC2pred_best_rank            2.818  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "64                    MixMHC2pred_best_allele       DRB1_01_01  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "66                      vaxrank_binding_score           0.0322  jETwpX0R9iEiQz2SMpHkPQ==\n",
       "67                        vaxrank_total_score          0.01434  jETwpX0R9iEiQz2SMpHkPQ=="
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations_ts"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
